diff --git a/Source/indra/newview/desktopnotifierlinux.cpp b/Source/indra/newview/desktopnotifierlinux.cpp
index a619c34..f3b2508 100644
--- a/Source/indra/newview/desktopnotifierlinux.cpp
+++ b/Source/indra/newview/desktopnotifierlinux.cpp
@@ -284,6 +284,9 @@ void DesktopNotifierLinux::showNotification( const std::string& notification_tit
 	//LL_INFOS( "DesktopNotifierLinux" ) << "New notification message: " << notification_message << LL_ENDL;
 	//LL_INFOS( "DesktopNotifierLinux" ) << "New notification type: " << notification_type << LL_ENDL;
 
+	if( !isUsable() )
+		return;
+
 	m_pLibNotify->mNotificationUpdate( m_pNotification,(gchar*)notification_title.c_str(), (gchar*)notification_message.c_str(), m_strIcon.c_str() );
 
 	m_pLibNotify->mNotificationSetUrgency( m_pNotification, NOTIFY_URGENCY_LOW );
@@ -293,11 +296,11 @@ void DesktopNotifierLinux::showNotification( const std::string& notification_tit
 	GError* error(0);
 	if ( m_pLibNotify->mNotificationShow( m_pNotification, &error ) )
 	{
-		LL_INFOS( "DesktopNotifierLinux" ) << "Linux desktop notification type " << notification_type << "sent." << LL_ENDL;
+		LL_INFOS( "DesktopNotifierLinux" ) << "Linux desktop notification type " << notification_type << " sent." << LL_ENDL;
 	}
 	else
 	{
-		LL_WARNS( "DesktopNotifierLinux" ) << "Linux desktop notification FAILED to send. " << error->message << LL_ENDL;
+		LL_WARNS( "DesktopNotifierLinux" ) << "Linux desktop notification type " << notification_type << " FAILED to send, error was " << error->message << LL_ENDL;
 	}
 }
 
