diff --git a/Source/indra/newview/fspanelclassified.h b/Source/indra/newview/fspanelclassified.h
index e61506c..286cfe9 100644
--- a/Source/indra/newview/fspanelclassified.h
+++ b/Source/indra/newview/fspanelclassified.h
@@ -27,8 +27,9 @@
 // Display of a classified used both for the global view in the
 // Find directory, and also for each individual user's classified in their
 // profile.
-#ifndef LL_FSPANELCLASSIFIED_H
-#define LL_FSPANELCLASSIFIED_H
+
+#ifndef FS_PANELCLASSIFIED_H
+#define FS_PANELCLASSIFIED_H
 
 #include "llavatarpropertiesprocessor.h"
 #include "llclassifiedinfo.h"
@@ -284,6 +285,7 @@ protected:
 	void onTextureSelected();
 
 private:
+	S32 getClassifiedFee();	// <FS:CR> FIRE-9814 - Don't hardcode a classified listing fee
 	bool mIsNew;
 	bool mIsNewWithErrors;
 	bool mCanClose;
@@ -293,4 +295,4 @@ private:
 	commit_signal_t mSaveButtonClickedSignal;
 };
 
-#endif // LL_FSPANELCLASSIFIED_H
+#endif // FS_PANELCLASSIFIED_H
