diff --git a/Source/indra/newview/lggbeamcolormapfloater.h b/Source/indra/newview/lggbeamcolormapfloater.h
index 47aceb1..7aa5d3c 100644
--- a/Source/indra/newview/lggbeamcolormapfloater.h
+++ b/Source/indra/newview/lggbeamcolormapfloater.h
@@ -11,6 +11,8 @@
  * http://sam.zoy.org/wtfpl/COPYING for more details.
  */
 
+#ifndef LGG_BEAMCOLORMAPFLOATER_H
+#define LGG_BEAMCOLORMAPFLOATER_H
 
 #include "llviewerprecompiledheaders.h"
 
@@ -70,3 +72,5 @@ protected:
 	LLSliderCtrl* mColorSlider;
 
 };
+
+#endif // LGG_BEAMCOLORMAPFLOATER_H
\ No newline at end of file
