diff --git a/Source/indra/newview/lggbeammaps.h b/Source/indra/newview/lggbeammaps.h
index 59ef427..60fc749 100644
--- a/Source/indra/newview/lggbeammaps.h
+++ b/Source/indra/newview/lggbeammaps.h
@@ -11,6 +11,9 @@
  * http://sam.zoy.org/wtfpl/COPYING for more details.
  */
 
+#ifndef LGG_BEAMMAPS_H
+#define LGG_BEAMMAPS_H
+
 #include "llhudeffecttrail.h"
 #include "llviewerprecompiledheaders.h"
 #include "lggbeamscolors.h"
@@ -51,3 +54,5 @@ class lggBeamMaps
 
 
 extern lggBeamMaps gLggBeamMaps;
+
+#endif // LGG_BEAMMAPS_H
\ No newline at end of file
