diff --git a/Source/indra/newview/llaccountingcostmanager.cpp b/Source/indra/newview/llaccountingcostmanager.cpp
index 8767955..7662a96 100644
--- a/Source/indra/newview/llaccountingcostmanager.cpp
+++ b/Source/indra/newview/llaccountingcostmanager.cpp
@@ -56,9 +56,9 @@ public:
 		}
 	}
 	
-	void error( U32 statusNum, const std::string& reason )
+	void errorWithContent( U32 statusNum, const std::string& reason, const LLSD& content )
 	{
-		llwarns	<< "Transport error "<<reason<<llendl;	
+		llwarns << "Transport error [status:" << statusNum << "]: " << content <<llendl;
 		clearPendingRequests();
 
 		LLAccountingCostObserver* observer = mObserverHandle.get();
