diff --git a/Source/indra/newview/llappviewer.cpp b/Source/indra/newview/llappviewer.cpp
index 91bfa4a..db6d19b 100644
--- a/Source/indra/newview/llappviewer.cpp
+++ b/Source/indra/newview/llappviewer.cpp
@@ -55,6 +55,7 @@
 #include "llstartup.h"
 #include "llfocusmgr.h"
 #include "llviewerjoystick.h"
+#include "llviewerremotecontrol.h"
 #include "llallocator.h"
 #include "llares.h" 
 #include "llcurl.h"
@@ -1335,6 +1336,7 @@ bool LLAppViewer::mainLoop()
 	LLTimer frameTimer,idleTimer;
 	LLTimer debugTime;
 	LLViewerJoystick* joystick(LLViewerJoystick::getInstance());
+	LLViewerRemoteControl* remoteControl(LLViewerRemoteControl::getInstance());
 	joystick->setNeedsReset(true);
 
     LLEventPump& mainloop(LLEventPumps::instance().obtain("mainloop"));
@@ -1404,6 +1406,7 @@ bool LLAppViewer::mainLoop()
 			if (!LLApp::isExiting())
 			{
 				pingMainloopTimeout("Main:JoystickKeyboard");
+				LLFollowCamMgr::updateScriptFollowCams();
 				
 				// Scan keyboard for movement keys.  Command keys and typing
 				// are handled by windows callbacks.  Don't do this until we're
@@ -1417,6 +1420,7 @@ bool LLAppViewer::mainLoop()
 				{
 					LLMemType mjk(LLMemType::MTYPE_JOY_KEY);
 					joystick->scanJoystick();
+					remoteControl->Tick();
 					// <FS:Ansariel> Chalice Yao's crouch toggle
 					static LLCachedControl<bool> fsCrouchToggle(gSavedSettings, "FSCrouchToggle");
 					static LLCachedControl<bool> fsCrouchToggleStatus(gSavedSettings, "FSCrouchToggleStatus");
