diff --git a/Source/indra/newview/llassetuploadresponders.h b/Source/indra/newview/llassetuploadresponders.h
index 1069b1b..224c6b8 100644
--- a/Source/indra/newview/llassetuploadresponders.h
+++ b/Source/indra/newview/llassetuploadresponders.h
@@ -42,7 +42,7 @@ public:
 							LLAssetType::EType asset_type);
 	~LLAssetUploadResponder();
 
-    virtual void error(U32 statusNum, const std::string& reason);
+    virtual void errorWithContent(U32 statusNum, const std::string& reason, const LLSD& content);
 	virtual void result(const LLSD& content);
 	virtual void uploadUpload(const LLSD& content);
 	virtual void uploadComplete(const LLSD& content);
@@ -67,7 +67,7 @@ public:
 		const LLSD& post_data,
 		const std::string& file_name,
 		LLAssetType::EType asset_type);
-    virtual void error(U32 statusNum, const std::string& reason);
+    virtual void errorWithContent(U32 statusNum, const std::string& reason, const LLSD& content);
 	virtual void uploadComplete(const LLSD& content);
 	virtual void uploadFailure(const LLSD& content);
 };
@@ -122,7 +122,7 @@ public:
 	~LLSendTexLayerResponder();
 
 	virtual void uploadComplete(const LLSD& content);
-	virtual void error(U32 statusNum, const std::string& reason);
+	virtual void errorWithContent(U32 statusNum, const std::string& reason, const LLSD& content);
 
 	LLBakedUploadData * mBakedUploadData;
 };
