diff --git a/Source/indra/newview/llaudiosourcevo.cpp b/Source/indra/newview/llaudiosourcevo.cpp
index b37aba6..2438193 100644
--- a/Source/indra/newview/llaudiosourcevo.cpp
+++ b/Source/indra/newview/llaudiosourcevo.cpp
@@ -34,7 +34,10 @@
 #include "llviewerparcelmgr.h"
 
 LLAudioSourceVO::LLAudioSourceVO(const LLUUID &sound_id, const LLUUID& owner_id, const F32 gain, LLViewerObject *objectp)
-	:	LLAudioSource(sound_id, owner_id, gain, LLAudioEngine::AUDIO_TYPE_SFX), 
+// <FS:CR> FIRE-10512 - Sound explorer fix by Sei Lisa
+//	:	LLAudioSource(sound_id, owner_id, gain, LLAudioEngine::AUDIO_TYPE_SFX),
+	:	LLAudioSource(sound_id, owner_id, gain, LLAudioEngine::AUDIO_TYPE_SFX, objectp->getID(), false),
+// </FS:CR>
 	mObjectp(objectp)
 {
 	update();
