diff --git a/Source/indra/newview/llavataractions.h b/Source/indra/newview/llavataractions.h
index e7ac7bc..30f0115 100644
--- a/Source/indra/newview/llavataractions.h
+++ b/Source/indra/newview/llavataractions.h
@@ -36,6 +36,7 @@
 
 class LLInventoryPanel;
 class LLFloater;
+class LLAvatarName;
 
 /**
  * Friend-related actions (add, remove, offer teleport, etc)
@@ -239,11 +240,24 @@ public:
 	static void estateKickMultiple(const uuid_vec_t& idAgents);
 	static void estateTeleportHomeMultiple(const uuid_vec_t& idAgents);
 
+	// <FS:Ansariel> Estate ban
+	static void estateBan(const LLUUID& idAgent);
+	static void estateBanMultiple(const uuid_vec_t& idAgents);
+	// </FS:Ansariel> Estate ban
+
+	// <FS:Ansariel> Derender
+	static void derender(const LLUUID& agent_id, bool permanent);
+	static void derenderMultiple(const uuid_vec_t& agent_ids, bool permanent);
+	static void onDerenderAvatarNameLookup(const LLUUID& agent_id, const LLAvatarName& av_name, bool permanent);
+	// </FS:Ansariel> Derender
+
 	static std::set<LLUUID> getInventorySelectedUUIDs();
 
 protected:
 	static bool callbackEstateKick(const LLSD& notification, const LLSD& response);
 	static bool callbackEstateTeleportHome(const LLSD& notification, const LLSD& response);
+	// <FS:Ansariel> Estate ban
+	static bool callbackEstateBan(const LLSD& notification, const LLSD& response);
 	
 private:
 	static bool callbackAddFriendWithMessage(const LLSD& notification, const LLSD& response);
