diff --git a/Source/indra/newview/llavatarlistitem.cpp b/Source/indra/newview/llavatarlistitem.cpp
index cedfe61..e4c19ba 100644
--- a/Source/indra/newview/llavatarlistitem.cpp
+++ b/Source/indra/newview/llavatarlistitem.cpp
@@ -95,7 +95,7 @@ LLAvatarListItem::LLAvatarListItem(bool not_from_ui_factory/* = true*/)
 	mShowPermissions(false),
 	mHovered(false),
 	mShowVoiceVolume(false),
-	mNearbyRange(false),
+	mNearbyRange(0),
 	mShowDisplayName(true),
 	mShowUsername(true),
 	mFirstSeen(time(NULL)),
@@ -220,13 +220,12 @@ S32 LLAvatarListItem::notifyParent(const LLSD& info)
 void LLAvatarListItem::onMouseEnter(S32 x, S32 y, MASK mask)
 {
 	getChildView("hovered_icon")->setVisible( true);
-
 	// AO, removed on-hover visibility. Don't do this. instead flip info buttons on full-time in postbuild.
 //	mInfoBtn->setVisible(mShowInfoBtn);
 //	mProfileBtn->setVisible(mShowProfileBtn);
 // [RLVa:KB] - Checked: 2010-04-05 (RLVa-1.2.2a) | Added: RLVa-1.2.0d
-//	mInfoBtn->setVisible( (mShowInfoBtn) && ((!mRlvCheckShowNames) || (!gRlvHandler.hasBehaviour(RLV_BHVR_SHOWNAMES))) );
-//	mProfileBtn->setVisible( (mShowProfileBtn) && ((!mRlvCheckShowNames) || (!gRlvHandler.hasBehaviour(RLV_BHVR_SHOWNAMES))) );
+	//mInfoBtn->setVisible( (mShowInfoBtn) && ((!mRlvCheckShowNames) || (!gRlvHandler.hasBehaviour(RLV_BHVR_SHOWNAMES))) );
+	//mProfileBtn->setVisible( (mShowProfileBtn) && ((!mRlvCheckShowNames) || (!gRlvHandler.hasBehaviour(RLV_BHVR_SHOWNAMES))) );
 // [/RLVa:KB]
 
 	mHovered = true;
@@ -715,7 +714,7 @@ void LLAvatarListItem::onAvatarNameCache(const LLAvatarName& av_name)
 
 	//KC - store the username for use in sorting
 	mUserName = av_name.mUsername;
-
+	
 	//requesting the list to resort
 	notifyParent(LLSD().with("sort", LLSD()));
 	
