diff --git a/Source/indra/newview/llchathistory.h b/Source/indra/newview/llchathistory.h
index e85d091..e396c70 100644
--- a/Source/indra/newview/llchathistory.h
+++ b/Source/indra/newview/llchathistory.h
@@ -24,13 +24,14 @@
  * $/LicenseInfo$
  */
 
+#if 0
+
 #ifndef LLCHATHISTORY_H_
 #define LLCHATHISTORY_H_
 
 #include "lltexteditor.h"
 #include "lltextbox.h"
 #include "llviewerchat.h"
-#include "llavatarname.h"
 
 //Chat log widget allowing addition of a message as a widget 
 class LLChatHistory : public LLUICtrl
@@ -95,8 +96,6 @@ class LLChatHistory : public LLUICtrl
 		 * Builds a message header.
 		 * @return pointer to LLView header object.
 		 */
-
-		void onAvatarNameCache(const LLUUID& agent_id, const LLAvatarName& av_name);
 		LLView* getHeader(const LLChat& chat,const LLStyle::Params& style_params, const LLSD& args);
 
 		void onClickMoreText();
@@ -140,9 +139,6 @@ class LLChatHistory : public LLUICtrl
 		S32 mBottomSeparatorPad;
 		S32 mTopHeaderPad;
 		S32 mBottomHeaderPad;
-	
-		std::string mDisplayName;
-		std::string mDisplayName_Username;
 
 		class LLLayoutPanel*	mMoreChatPanel;
 		LLTextBox*		mMoreChatText;
@@ -151,3 +147,5 @@ class LLChatHistory : public LLUICtrl
 		unread_chat_source_t mUnreadChatSources;
 };
 #endif /* LLCHATHISTORY_H_ */
+
+#endif
