diff --git a/Source/indra/newview/llclassifiedstatsresponder.h b/Source/indra/newview/llclassifiedstatsresponder.h
index 3db1868..06dcb62 100644
--- a/Source/indra/newview/llclassifiedstatsresponder.h
+++ b/Source/indra/newview/llclassifiedstatsresponder.h
@@ -39,7 +39,7 @@ public:
 	virtual void result(const LLSD& content);
 	//If we get back an error (not found, etc...), handle it here
 	
-	virtual void error(U32 status, const std::string& reason);
+	virtual void errorWithContent(U32 status, const std::string& reason, const LLSD& content);
 
 protected:
 	LLUUID mClassifiedID;
