diff --git a/Source/indra/newview/llcompilequeue.cpp b/Source/indra/newview/llcompilequeue.cpp
index 6c26b4f..a62017b 100644
--- a/Source/indra/newview/llcompilequeue.cpp
+++ b/Source/indra/newview/llcompilequeue.cpp
@@ -165,18 +165,6 @@ BOOL LLFloaterScriptQueue::start()
 	//llinfos << "LLFloaterCompileQueue::start()" << llendl;
 	std::string buffer;
 
-	LLSelectMgr *mgr = LLSelectMgr::getInstance();
-	LLObjectSelectionHandle selectHandle = mgr->getSelection();
-	U32 n_objects = 0;
-	if (gSavedSettings.getBOOL("EditLinkedParts"))
-	{
-		n_objects = selectHandle->getObjectCount();
-	}
-	else
-	{
-		n_objects = selectHandle->getRootObjectCount();
-	}
-
 	LLStringUtil::format_map_t args;
 	args["[START]"] = mStartString;
 	args["[COUNT]"] = llformat ("%d", mObjectIDs.count());
@@ -196,18 +184,6 @@ bool LLFloaterScriptQueue::onScriptModifyConfirmation(const LLSD& notification,
 	}
 	std::string buffer;
 
-	LLSelectMgr *mgr = LLSelectMgr::getInstance();
-	LLObjectSelectionHandle selectHandle = mgr->getSelection();
-	U32 n_objects = 0;
-	if (gSavedSettings.getBOOL("EditLinkedParts"))
-	{
-		n_objects = selectHandle->getObjectCount();
-	}
-	else
-	{
-		n_objects = selectHandle->getRootObjectCount();
-	}
-
 	LLStringUtil::format_map_t args;
 	args["[START]"] = mStartString;
 	args["[COUNT]"] = llformat ("%d", mObjectIDs.count());
