diff --git a/Source/indra/newview/llexternaleditor.cpp b/Source/indra/newview/llexternaleditor.cpp
index 9480e54..0560775 100644
--- a/Source/indra/newview/llexternaleditor.cpp
+++ b/Source/indra/newview/llexternaleditor.cpp
@@ -45,7 +45,25 @@ LLExternalEditor::EErrorCode LLExternalEditor::setCommand(const std::string& env
 	if (cmd.empty())
 	{
 		llwarns << "Editor command is empty or not set" << llendl;
-		return EC_NOT_SPECIFIED;
+// <FS:CR> FIRE-10320 If no editor is set, fallback on the system open handler
+		//return EC_NOT_SPECIFIED;
+		llwarns << "Falling back on generic open handler" << llendl;
+#if LL_WINDOWS
+		std::string comspec(getenv("COMSPEC"));
+		comspec.append(" /C START \"%s\"");
+		cmd = findCommand("", comspec);
+#elif LL_DARWIN
+		cmd = findCommand("", "/usr/bin/open \"%s\"");
+#elif LL_LINUX
+		// xdg-open might not actually be installed, but it's out best shot
+		cmd = findCommand("", "/usr/bin/xdg-open \"%s\"");
+#endif
+		if (cmd.empty())
+		{
+			llwarns << "Failed to find generic open handler: " << cmd << llendl;
+			return EC_NOT_SPECIFIED;
+		}
+// </FS:CR>
 	}
 
 	string_vec_t tokens;
