diff --git a/Source/indra/newview/llfeaturemanager.h b/Source/indra/newview/llfeaturemanager.h
index c9cb397..ad72c16 100644
--- a/Source/indra/newview/llfeaturemanager.h
+++ b/Source/indra/newview/llfeaturemanager.h
@@ -39,7 +39,9 @@ typedef enum EGPUClass
 	GPU_CLASS_0 = 0,
 	GPU_CLASS_1 = 1,
 	GPU_CLASS_2 = 2,
-	GPU_CLASS_3 = 3
+	GPU_CLASS_3 = 3,
+	GPU_CLASS_4 = 4,
+	GPU_CLASS_5 = 5
 } EGPUClass; 
 
 
@@ -101,7 +103,8 @@ public:
 		mTableVersion(0),
 		mSafe(FALSE),
 		mGPUClass(GPU_CLASS_UNKNOWN),
-		mGPUSupported(FALSE)
+		mExpectedGLVersion(0.f),
+		mGPUSupported(FALSE)		
 	{
 	}
 	~LLFeatureManager() {cleanupFeatureTables();}
@@ -116,6 +119,7 @@ public:
 	EGPUClass getGPUClass() 			{ return mGPUClass; }
 	std::string& getGPUString() 		{ return mGPUString; }
 	BOOL isGPUSupported()				{ return mGPUSupported; }
+	F32 getExpectedGLVersion()			{ return mExpectedGLVersion; }
 	
 	void cleanupFeatureTables();
 
@@ -155,6 +159,7 @@ protected:
 	S32			mTableVersion;
 	BOOL		mSafe;					// Reinitialize everything to the "safe" mask
 	EGPUClass	mGPUClass;
+	F32			mExpectedGLVersion;		//expected GL version according to gpu table
 	std::string	mGPUString;
 	BOOL		mGPUSupported;
 };
