diff --git a/Source/indra/newview/llfloaterscriptlimits.cpp b/Source/indra/newview/llfloaterscriptlimits.cpp
index 6139a83..4d3c438 100644
--- a/Source/indra/newview/llfloaterscriptlimits.cpp
+++ b/Source/indra/newview/llfloaterscriptlimits.cpp
@@ -51,6 +51,8 @@
 #include "llviewerregion.h"
 #include "llviewerwindow.h"
 
+#include "fslslbridge.h"
+
 ///----------------------------------------------------------------------------
 /// LLFloaterScriptLimits
 ///----------------------------------------------------------------------------
@@ -221,9 +223,9 @@ void fetchScriptLimitsRegionInfoResponder::result(const LLSD& content)
 	}
 }
 
-void fetchScriptLimitsRegionInfoResponder::error(U32 status, const std::string& reason)
+void fetchScriptLimitsRegionInfoResponder::errorWithContent(U32 status, const std::string& reason, const LLSD& content)
 {
-	llwarns << "Error from responder " << reason << llendl;
+	llwarns << "fetchScriptLimitsRegionInfoResponder error [status:" << status << "]: " << content << llendl;
 }
 
 void fetchScriptLimitsRegionSummaryResponder::result(const LLSD& content_ref)
@@ -308,9 +310,9 @@ void fetchScriptLimitsRegionSummaryResponder::result(const LLSD& content_ref)
 	}
 }
 
-void fetchScriptLimitsRegionSummaryResponder::error(U32 status, const std::string& reason)
+void fetchScriptLimitsRegionSummaryResponder::errorWithContent(U32 status, const std::string& reason, const LLSD& content)
 {
-	llwarns << "Error from responder " << reason << llendl;
+	llwarns << "fetchScriptLimitsRegionSummaryResponder error [status:" << status << "]: " << content << llendl;
 }
 
 void fetchScriptLimitsRegionDetailsResponder::result(const LLSD& content_ref)
@@ -417,9 +419,9 @@ result (map)
 	}
 }
 
-void fetchScriptLimitsRegionDetailsResponder::error(U32 status, const std::string& reason)
+void fetchScriptLimitsRegionDetailsResponder::errorWithContent(U32 status, const std::string& reason, const LLSD& content)
 {
-	llwarns << "Error from responder " << reason << llendl;
+	llwarns << "fetchScriptLimitsRegionDetailsResponder error [status:" << status << "]: " << content << llendl;
 }
 
 void fetchScriptLimitsAttachmentInfoResponder::result(const LLSD& content_ref)
@@ -513,9 +515,9 @@ void fetchScriptLimitsAttachmentInfoResponder::result(const LLSD& content_ref)
 	}
 }
 
-void fetchScriptLimitsAttachmentInfoResponder::error(U32 status, const std::string& reason)
+void fetchScriptLimitsAttachmentInfoResponder::errorWithContent(U32 status, const std::string& reason, const LLSD& content)
 {
-	llwarns << "Error from responder " << reason << llendl;
+	llwarns << "fetchScriptLimitsAttachmentInfoResponder error [status:" << status << "]: " << content << llendl;
 }
 
 ///----------------------------------------------------------------------------
@@ -1189,11 +1191,17 @@ void LLPanelScriptLimitsAttachment::setAttachmentDetails(LLSD content)
 
 	for(int i = 0; i < number_attachments; i++)
 	{
+		// <FS:Ansariel> Firestorm Bridge hack
+		bool isValid = true;
+
 		std::string humanReadableLocation = "";
 		if(content["attachments"][i].has("location"))
 		{
 			std::string actualLocation = content["attachments"][i]["location"];
 			humanReadableLocation = LLTrans::getString(actualLocation.c_str());
+
+			// <FS:Ansariel> Firestorm Bridge hack
+			isValid = (actualLocation != "Invalid Attachment");
 		}
 		
 		S32 number_objects = content["attachments"][i]["objects"].size();
@@ -1212,13 +1220,13 @@ void LLPanelScriptLimitsAttachment::setAttachmentDetails(LLSD content)
 			}
 			std::string name = content["attachments"][i]["objects"][j]["name"].asString();
 
-			// Ansariel: Crude hack to make the bridge display the proper attachment spot
-			//           and not "MissingString(Bad attachment point)"
-			if (humanReadableLocation == "MissingString(Bad attachment point)" && (name.find("Firestorm LSL Bridge") != std::string::npos || name == "Rock - medium, round"))
+			// <FS:Ansariel> Firestorm Bridge hack
+			if (!isValid && (name.find(FS_BRIDGE_NAME) != std::string::npos || name == LIB_ROCK_NAME))
 			{
 				humanReadableLocation = LLTrans::getString("Bridge");
 				name = "Firestorm LSL Bridge";
 			}
+			// </FS:Ansariel>
 			
 			LLSD element;
 
