diff --git a/Source/indra/newview/llfloaterscriptdebug.cpp b/Source/indra/newview/llfloaterscriptdebug.cpp
index b691db1..744289f 100644
--- a/Source/indra/newview/llfloaterscriptdebug.cpp
+++ b/Source/indra/newview/llfloaterscriptdebug.cpp
@@ -65,13 +65,21 @@ LLFloaterScriptDebug::~LLFloaterScriptDebug()
 
 void LLFloaterScriptDebug::show(const LLUUID& object_id)
 {
-	addOutputWindow(object_id);
+	// <FS:Ansariel> Script debug icon
+	//addOutputWindow(object_id);
+	addOutputWindow(object_id, true);
+	// </FS:Ansariel> Script debug icon
 }
 
 BOOL LLFloaterScriptDebug::postBuild()
 {
 	LLMultiFloater::postBuild();
 
+	// <FS:Ansariel> Disconnect close signal. It would call closeAllFloaters()
+	//               and the floaters in the tabs will be gone since LL
+	//               messed with the floater instance handling!
+	mCloseSignal.disconnect_all_slots();
+
 	if (mTabContainer)
 	{
 		return TRUE;
@@ -80,7 +88,10 @@ BOOL LLFloaterScriptDebug::postBuild()
 	return FALSE;
 }
 
-LLFloater* LLFloaterScriptDebug::addOutputWindow(const LLUUID &object_id)
+// <FS:Ansariel> Script debug icon
+//LLFloater* LLFloaterScriptDebug::addOutputWindow(const LLUUID &object_id)
+LLFloater* LLFloaterScriptDebug::addOutputWindow(const LLUUID& object_id, bool show /* = false */)
+// </FS:Ansariel> Script debug icon
 {
 	LLMultiFloater* host = LLFloaterReg::showTypedInstance<LLMultiFloater>("script_debug", LLSD());
 	if (!host)
@@ -91,6 +102,13 @@ LLFloater* LLFloaterScriptDebug::addOutputWindow(const LLUUID &object_id)
 	LLFloater* floaterp = LLFloaterReg::showInstance("script_debug_output", object_id, FALSE);
 	LLFloater::setFloaterHost(NULL);
 
+	// <FS:Ansariel> Script debug icon
+	if (gSavedSettings.getS32("ShowScriptErrorsLocation") == 0 && !show)
+	{
+		host->closeFloater();
+	}
+	// </FS:Ansariel> Script debug icon
+
 	return floaterp;
 }
 
@@ -105,8 +123,11 @@ void LLFloaterScriptDebug::addScriptLine(const std::string &utf8mesg, const std:
 
 	if (objectp)
 	{
-		objectp->setIcon(LLViewerTextureManager::getFetchedTextureFromFile("script_error.j2c", TRUE, LLViewerTexture::BOOST_UI));
-		floater_label = llformat("%s(%.0f, %.0f, %.0f)",
+		objectp->setIcon(LLViewerTextureManager::getFetchedTextureFromFile("script_error.j2c", FTT_LOCAL_FILE, TRUE, LLGLTexture::BOOST_UI));
+		// <FS:Ansariel> Mark script error icons
+		objectp->getIcon()->setScriptError();
+		// </FS:Ansariel> Mark script error icons
+		floater_label = llformat("%s (%.0f, %.0f, %.0f)",
 						user_name.c_str(),
 						objectp->getPositionRegion().mV[VX],
 						objectp->getPositionRegion().mV[VY],
