diff --git a/Source/indra/newview/llfloatersettingsdebug.cpp b/Source/indra/newview/llfloatersettingsdebug.cpp
index f2fa5d9..0cbc8cb 100644
--- a/Source/indra/newview/llfloatersettingsdebug.cpp
+++ b/Source/indra/newview/llfloatersettingsdebug.cpp
@@ -337,6 +337,7 @@ void LLFloaterSettingsDebug::updateControl()
 	{
 // [RLVa:KB] - Checked: 2011-05-28 (RLVa-1.4.0a) | Modified: RLVa-1.4.0a
 		// If "HideFromEditor" was toggled while the floater is open then we need to manually disable access to the control
+		// NOTE: this runs per-frame so there's no need to explictly handle onCommitSettings() or onClickDefault()
 		mOldVisibility=mCurrentControlVariable->isHiddenFromSettingsEditor();
 		mSpinner1->setEnabled(!mOldVisibility);
 		mSpinner2->setEnabled(!mOldVisibility);
@@ -348,6 +349,7 @@ void LLFloaterSettingsDebug::updateControl()
 		mDefaultButton->setEnabled(!mOldVisibility);
 // [/RLVa:KB]
 
+
 		mCopyButton->setEnabled(TRUE);
 		mSanityButton->setVisible(!mCurrentControlVariable->isSane());
 
