diff --git a/Source/indra/newview/llfloaterurlentry.cpp b/Source/indra/newview/llfloaterurlentry.cpp
index 151cd2a..e85d849 100644
--- a/Source/indra/newview/llfloaterurlentry.cpp
+++ b/Source/indra/newview/llfloaterurlentry.cpp
@@ -219,7 +219,8 @@ void LLFloaterURLEntry::onBtnOK( void* userdata )
 	}
 
 	// Discover the MIME type only for "http" scheme.
-	if(scheme == "http" || scheme == "https")
+	if(!media_url.empty() && 
+	   (scheme == "http" || scheme == "https"))
 	{
 		LLHTTPClient::getHeaderOnly( media_url,
 			new LLMediaTypeResponder(self->getHandle()));
