diff --git a/Source/indra/newview/llfloatergesture.cpp b/Source/indra/newview/llfloatergesture.cpp
index 40c42c7..97db1e1 100644
--- a/Source/indra/newview/llfloatergesture.cpp
+++ b/Source/indra/newview/llfloatergesture.cpp
@@ -65,40 +65,27 @@ private:
 	LLFloaterGesture* mFloater;
 };
 //-----------------------------
-// GestureCallback
+// gesture callback funcs
 //-----------------------------
 
-class GestureShowCallback : public LLInventoryCallback
+void gesture_show_cb(const LLUUID &inv_item)
 {
-public:
-	void fire(const LLUUID &inv_item)
-	{
-		LLPreviewGesture::show(inv_item, LLUUID::null);
-	}
-};
+	LLPreviewGesture::show(inv_item, LLUUID::null);
+}
 
-class GestureCopiedCallback : public LLInventoryCallback
+void gesture_copied_cb(const LLUUID &inv_item, LLFloaterGesture* floater)
 {
-private:
-	LLFloaterGesture* mFloater;
-	
-public:
-	GestureCopiedCallback(LLFloaterGesture* floater): mFloater(floater)
-	{}
-	void fire(const LLUUID &inv_item)
+	if(floater)
 	{
-		if(mFloater)
-		{
-			mFloater->addGesture(inv_item,NULL,mFloater->getChild<LLScrollListCtrl>("gesture_list"));
-
-			// EXP-1909 (Pasted gesture displayed twice)
-			// The problem is that addGesture is called here for the second time for the same item (which is copied)
-			// First time addGesture is called from LLFloaterGestureObserver::changed(), which is a callback for inventory
-			// change. So we need to refresh the gesture list to avoid duplicates.
-			mFloater->refreshAll();
-		}
+		floater->addGesture(inv_item,NULL,floater->getChild<LLScrollListCtrl>("gesture_list"));
+		
+		// EXP-1909 (Pasted gesture displayed twice)
+		// The problem is that addGesture is called here for the second time for the same item (which is copied)
+		// First time addGesture is called from LLFloaterGestureObserver::changed(), which is a callback for inventory
+		// change. So we need to refresh the gesture list to avoid duplicates.
+		floater->refreshAll();
 	}
-};
+}
 
 //---------------------------------------------------------------------------
 // LLFloaterGesture
@@ -470,7 +457,7 @@ void LLFloaterGesture::onClickPlay()
 
 void LLFloaterGesture::onClickNew()
 {
-	LLPointer<LLInventoryCallback> cb = new GestureShowCallback();
+	LLPointer<LLInventoryCallback> cb = new LLBoostFuncInventoryCallback(gesture_show_cb);
 	create_inventory_item(gAgent.getID(), gAgent.getSessionID(),
 		LLUUID::null, LLTransactionID::tnull, "New Gesture", "", LLAssetType::AT_GESTURE,
 		LLInventoryType::IT_GESTURE, NOT_WEARABLE, PERM_MOVE | PERM_TRANSFER, cb);
@@ -542,7 +529,7 @@ void LLFloaterGesture::onCopyPasteAction(const LLSD& command)
 			return;
 		LLInventoryCategory* gesture_dir = gInventory.getCategory(mGestureFolderID);
 		llassert(gesture_dir);
-		LLPointer<GestureCopiedCallback> cb = new GestureCopiedCallback(this);
+		LLPointer<LLInventoryCallback> cb = new LLBoostFuncInventoryCallback(boost::bind(gesture_copied_cb, _1, this));
 
 		for(LLDynamicArray<LLUUID>::iterator it = ids.begin(); it != ids.end(); it++)
 		{
