diff --git a/Source/indra/newview/llfloaterbuycontents.cpp b/Source/indra/newview/llfloaterbuycontents.cpp
index bca4b5e..aa6ace2 100644
--- a/Source/indra/newview/llfloaterbuycontents.cpp
+++ b/Source/indra/newview/llfloaterbuycontents.cpp
@@ -40,6 +40,7 @@
 #include "llcheckboxctrl.h"
 #include "llinventorydefines.h"
 #include "llinventoryfunctions.h"
+#include "llinventoryicon.h"
 #include "llinventorymodel.h"	// for gInventory
 #include "llfirstuse.h"
 #include "llfloaterreg.h"
@@ -80,6 +81,7 @@ BOOL LLFloaterBuyContents::postBuild()
 
 LLFloaterBuyContents::~LLFloaterBuyContents()
 {
+	removeVOInventoryListener();
 }
 
 
@@ -147,23 +149,26 @@ void LLFloaterBuyContents::inventoryChanged(LLViewerObject* obj,
 		return;
 	}
 
-	if (!inv)
+	LLScrollListCtrl* item_list = getChild<LLScrollListCtrl>("item_list");
+	if (!item_list)
 	{
-		llwarns << "No inventory in LLFloaterBuyContents::inventoryChanged"
-			<< llendl;
 		removeVOInventoryListener();
 		return;
 	}
 
-	LLCtrlListInterface *item_list = childGetListInterface("item_list");
-	if (!item_list)
+	item_list->deleteAllItems();
+	
+	if (!inv)
 	{
-		removeVOInventoryListener();
+		llwarns << "No inventory in LLFloaterBuyContents::inventoryChanged"
+			<< llendl;
+
 		return;
 	}
 
 	// default to turning off the buy button.
-	getChildView("buy_btn")->setEnabled(FALSE);
+	LLView* buy_btn = getChildView("buy_btn");
+	buy_btn->setEnabled(FALSE);
 
 	LLUUID owner_id;
 	BOOL is_group_owned;
@@ -204,7 +209,7 @@ void LLFloaterBuyContents::inventoryChanged(LLViewerObject* obj,
 
 		// There will be at least one item shown in the display, so go
 		// ahead and enable the buy button.
-		getChildView("buy_btn")->setEnabled(TRUE);
+		buy_btn->setEnabled(TRUE);
 
 		// Create the line in the list
 		LLSD row;
@@ -255,8 +260,6 @@ void LLFloaterBuyContents::inventoryChanged(LLViewerObject* obj,
 		getChildView("wear_check")->setEnabled(TRUE);
 		getChild<LLUICtrl>("wear_check")->setValue(LLSD(false) );
 	}
-	
-	removeVOInventoryListener();
 }
 
 
