diff --git a/Source/indra/newview/llfloaterregioninfo.h b/Source/indra/newview/llfloaterregioninfo.h
index 502e9e0..0d4699c 100644
--- a/Source/indra/newview/llfloaterregioninfo.h
+++ b/Source/indra/newview/llfloaterregioninfo.h
@@ -338,9 +338,6 @@ public:
 	const std::string getOwnerName() const;
 	void setOwnerName(const std::string& name);
 
-	// If visible from mainland, allowed agent and allowed groups
-	// are ignored, so must disable UI.
-	void setAccessAllowedEnabled(bool enable_agent, bool enable_group, bool enable_ban);
 protected:
 	virtual BOOL sendUpdate();
 	// confirmation dialog callback
@@ -350,7 +347,6 @@ protected:
 	void commitEstateManagers();
 	
 	void clearAccessLists();
-	BOOL checkRemovalButton(std::string name);
 	BOOL checkSunHourSlider(LLUICtrl* child_ctrl);
 
 	U32 mEstateID;
