diff --git a/Source/indra/newview/llfloatermodelpreview.h b/Source/indra/newview/llfloatermodelpreview.h
index c86c532..32ddad0 100644
--- a/Source/indra/newview/llfloatermodelpreview.h
+++ b/Source/indra/newview/llfloatermodelpreview.h
@@ -314,6 +314,15 @@ class LLModelPreview : public LLViewerDynamicTexture, public LLMutex
 	typedef boost::signals2::signal<void (bool)> model_updated_signal_t;
 
 public:
+
+	typedef enum
+	{
+		LOD_FROM_FILE = 0,
+		GENERATE,
+		USE_LOD_ABOVE,
+	} eLoDMode;
+
+public:
 	LLModelPreview(S32 width, S32 height, LLFloater* fmp);
 	virtual ~LLModelPreview();
 
