diff --git a/Source/indra/newview/llfolderview.cpp b/Source/indra/newview/llfolderview.cpp
index 4dfeb1c..50c7c5c 100644
--- a/Source/indra/newview/llfolderview.cpp
+++ b/Source/indra/newview/llfolderview.cpp
@@ -906,7 +906,6 @@ void LLFolderView::commitRename( const LLSD& data )
 
 void LLFolderView::draw()
 {
-	static LLUIColor sSearchStatusColor = LLUIColorTable::instance().getColor("InventorySearchStatusColor", LLColor4::white);
 	if (mDebugFilters)
 	{
 		std::string current_filter_string = llformat("Current Filter: %d, Least Filter: %d, Auto-accept Filter: %d",
@@ -1546,12 +1545,6 @@ BOOL LLFolderView::handleKeyHere( KEY key, MASK mask )
 		LLMenuGL::sMenuContainer->hideMenus();
 	}
 
-	LLView *item = NULL;
-	if (getChildCount() > 0)
-	{
-		item = *(getChildList()->begin());
-	}
-
 	switch( key )
 	{
 	case KEY_F2:
@@ -2079,7 +2072,6 @@ void LLFolderView::scrollToShowItem(LLFolderViewItem* item, const LLRect& constr
 	{
 		LLRect local_rect = item->getLocalRect();
 		LLRect item_scrolled_rect; // item position relative to display area of scroller
-		LLRect visible_doc_rect = mScrollContainer->getVisibleContentRect();
 		
 		S32 icon_height = mIcon.isNull() ? 0 : mIcon->getHeight(); 
 		S32 label_height = getLabelFontForStyle(mLabelStyle)->getLineHeight(); 
@@ -2279,9 +2271,16 @@ void LLFolderView::doIdle()
 		mDebugFilters = debug_filters;
 		arrangeAll();
 	}
-	BOOL filter_modified_and_active = mFilter->isModified() && mFilter->isNotDefault();
-	mNeedsAutoSelect = filter_modified_and_active &&
-						!(gFocusMgr.childHasKeyboardFocus(this) || gFocusMgr.getMouseCapture());
+//	BOOL filter_modified_and_active = mFilter->isModified() && mFilter->isNotDefault();
+//	mNeedsAutoSelect = filter_modified_and_active &&
+//						!(gFocusMgr.childHasKeyboardFocus(this) || gFocusMgr.getMouseCapture());
+
+// [SL:KB] - Patch: Inventory-Misc | Checked: 2013-05-02 (Catznip-3.4)
+	if (mFilter->isModified() && mFilter->isNotDefault())
+	{
+		mNeedsAutoSelect = TRUE;
+	}
+// [/SL:KB]
 	mFilter->clearModified();
 
 	// filter to determine visibility before arranging
