diff --git a/Source/indra/newview/llhomelocationresponder.h b/Source/indra/newview/llhomelocationresponder.h
index d640b9c..9bf4b12 100644
--- a/Source/indra/newview/llhomelocationresponder.h
+++ b/Source/indra/newview/llhomelocationresponder.h
@@ -36,7 +36,7 @@
 class LLHomeLocationResponder : public LLHTTPClient::Responder
 {
 	virtual void result( const LLSD& content );
-	virtual void error( U32 status, const std::string& reason );
+	virtual void errorWithContent( U32 status, const std::string& reason, const LLSD& content );
 };
 
 #endif
