diff --git a/Source/indra/newview/llimfloatercontainer.h b/Source/indra/newview/llimfloatercontainer.h
index 3c983a1..404f780 100644
--- a/Source/indra/newview/llimfloatercontainer.h
+++ b/Source/indra/newview/llimfloatercontainer.h
@@ -24,6 +24,8 @@
  * $/LicenseInfo$
  */
 
+#if 0
+
 #ifndef LL_LLIMFLOATERCONTAINER_H
 #define LL_LLIMFLOATERCONTAINER_H
 
@@ -46,13 +48,10 @@ public:
 	/*virtual*/ BOOL postBuild();
 	/*virtual*/ void onOpen(const LLSD& key);
 	void onCloseFloater(LLUUID& id);
-	
+
 	/*virtual*/ void addFloater(LLFloater* floaterp, 
 								BOOL select_added_floater, 
 								LLTabContainer::eInsertionPoint insertion_point = LLTabContainer::END);
-// [SL:KB] - Patch: Chat-NearbyChatBar | Checked: 2011-12-11 (Catznip-3.2.0d) | Added: Catznip-3.2.0d
-	/*virtual*/ void removeFloater(LLFloater* floaterp);
-// [/SL:KB]
 
 	static LLFloater* getCurrentVoiceFloater();
 
@@ -61,13 +60,15 @@ public:
 	static LLIMFloaterContainer* getInstance();
 
 	virtual void setMinimized(BOOL b);
-	
-	void onNewMessageReceived(const LLSD& data); // public so nearbychat can call it directly. TODO: handle via callback. -AO
 
 private:
 	typedef std::map<LLUUID,LLFloater*> avatarID_panel_map_t;
 	avatarID_panel_map_t mSessions;
 	boost::signals2::connection mNewMessageConnection;
+
+	void onNewMessageReceived(const LLSD& data);
 };
 
 #endif // LL_LLIMFLOATERCONTAINER_H
+
+#endif
