diff --git a/Source/indra/newview/llinventoryfunctions.h b/Source/indra/newview/llinventoryfunctions.h
index fb3b075..d26d381 100644
--- a/Source/indra/newview/llinventoryfunctions.h
+++ b/Source/indra/newview/llinventoryfunctions.h
@@ -60,6 +60,7 @@ void show_item_profile(const LLUUID& item_uuid);
 void show_task_item_profile(const LLUUID& item_uuid, const LLUUID& object_id);
 
 void show_item_original(const LLUUID& item_uuid);
+void reset_inventory_filter();
 
 void rename_category(LLInventoryModel* model, const LLUUID& cat_id, const std::string& new_name);
 
