diff --git a/Source/indra/newview/llinventorypanel.h b/Source/indra/newview/llinventorypanel.h
index 39e2a2e..4a71d00 100644
--- a/Source/indra/newview/llinventorypanel.h
+++ b/Source/indra/newview/llinventorypanel.h
@@ -245,16 +245,13 @@ public:
 	BOOL 				getIsViewsInitialized() const { return mViewsInitialized; }
 	const LLUUID&		getRootFolderID() const;
 
-	void	addSubItems(const LLUUID& id, long ); //ND Just-In-Time subfolder load
-
 protected:
 	// Builds the UI.  Call this once the inventory is usable.
 	void 				initializeViews();
 	LLFolderViewItem*	rebuildViewsFor(const LLUUID& id); // Given the id and the parent, build all of the folder views.
 
 	virtual void		buildFolderView(const LLInventoryPanel::Params& params);
-//	LLFolderViewItem*	buildNewViews(const LLUUID& id );
-	LLFolderViewItem*	buildNewViews(const LLUUID& id, long ); // <ND/> new arg to limit the depth of initial tree creation.
+	LLFolderViewItem*	buildNewViews(const LLUUID& id );
 	BOOL				getIsHiddenFolderType(LLFolderType::EType folder_type) const;
 	
 	virtual LLFolderView*		createFolderView(LLInvFVBridge * bridge, bool useLabelSuffix);
