diff --git a/Source/indra/newview/llinspectavatar.cpp b/Source/indra/newview/llinspectavatar.cpp
index 17d0b0f..aafc43b 100644
--- a/Source/indra/newview/llinspectavatar.cpp
+++ b/Source/indra/newview/llinspectavatar.cpp
@@ -501,9 +501,9 @@ void LLInspectAvatar::toggleSelectedVoice(bool enabled)
 				mSessionID = session_id;
 			}
 
-			virtual void error(U32 status, const std::string& reason)
+			virtual void errorWithContent(U32 status, const std::string& reason, const LLSD& content)
 			{
-				llwarns << status << ": " << reason << llendl;
+				llwarns << "MuteVoiceResponder error [status:" << status << "]: " << content << llendl;
 
 				if ( gIMMgr )
 				{
