diff --git a/Source/indra/newview/llmaniprotate.h b/Source/indra/newview/llmaniprotate.h
index e8f1c24..9295d28 100644
--- a/Source/indra/newview/llmaniprotate.h
+++ b/Source/indra/newview/llmaniprotate.h
@@ -87,7 +87,7 @@ private:
 	static LLVector3	intersectRayWithSphere( const LLVector3& ray_pt, const LLVector3& ray_dir, const LLVector3& sphere_center, F32 sphere_radius);
 
 private:
-	LLVector3d			mRotationCenter;			
+	LLVector3d			mRotationCenter;
 	LLCoordGL			mCenterScreen;
 //	S32					mLastHoverMouseX;		// used to suppress hover if mouse doesn't move
 //	S32					mLastHoverMouseY;
@@ -105,7 +105,9 @@ private:
 
 	BOOL				mSendUpdateOnMouseUp;
 
-	BOOL				mSmoothRotate;
+// <FS:CR> FIRE-8882
+	//BOOL				mSmoothRotate;
+// </FS:CR>
 	BOOL				mCamEdgeOn;
 
 	LLVector4			mManipulatorVertices[6];
