diff --git a/Source/indra/newview/llmanip.cpp b/Source/indra/newview/llmanip.cpp
index 26242c3..a4a9445 100644
--- a/Source/indra/newview/llmanip.cpp
+++ b/Source/indra/newview/llmanip.cpp
@@ -73,7 +73,6 @@ void LLManip::rebuild(LLViewerObject* vobj)
 	LLDrawable* drawablep = vobj->mDrawable;
 	if (drawablep && drawablep->getVOVolume())
 	{
-		
 		gPipeline.markRebuild(drawablep,LLDrawable::REBUILD_VOLUME, TRUE);
 		drawablep->setState(LLDrawable::MOVE_UNDAMPED); // force to UNDAMPED
 		drawablep->updateMove();
@@ -83,6 +82,14 @@ void LLManip::rebuild(LLViewerObject* vobj)
 			group->dirtyGeom();
 			gPipeline.markRebuild(group, TRUE);
 		}
+
+		LLViewerObject::const_child_list_t& child_list = vobj->getChildren();
+		for (LLViewerObject::child_list_t::const_iterator iter = child_list.begin(), endIter = child_list.end();
+			 iter != endIter; ++iter)
+		{
+			LLViewerObject* child = *iter;
+			rebuild(child);
+		}
 	}
 }
 
@@ -348,6 +355,13 @@ LLVector3 LLManip::getSavedPivotPoint() const
 
 LLVector3 LLManip::getPivotPoint()
 {
+	// <FS:KC> Pivot point controls
+	//if (mObjectSelection->getFirstObject() && mObjectSelection->getObjectCount() == 1 && mObjectSelection->getSelectType() != SELECT_TYPE_HUD)
+	//{
+	//	return mObjectSelection->getFirstObject()->getPivotPositionAgent();
+	//}
+	//return LLSelectMgr::getInstance()->getBBoxOfSelection().getCenterAgent();
+
 	LLVector3 pos;
 	LLVector3 scale;
 	LLQuaternion rot;
@@ -358,24 +372,26 @@ LLVector3 LLManip::getPivotPoint()
 	static LLCachedControl<F32> sPivotZ(gSavedSettings, "FSBuildPrefs_PivotZ");
 	
 	const BOOL children_ok = TRUE;
-	if (mObjectSelection->getFirstRootObject(children_ok) && (mObjectSelection->getObjectCount() == 1 || sActualRoot) && mObjectSelection->getSelectType() != SELECT_TYPE_HUD)
+	LLViewerObject* root_object = mObjectSelection->getFirstRootObject(children_ok);
+	if (root_object && (mObjectSelection->getObjectCount() == 1 || sActualRoot) && mObjectSelection->getSelectType() != SELECT_TYPE_HUD)
 	{
-		pos = mObjectSelection->getFirstRootObject(children_ok)->getPivotPositionAgent();
-		scale = mObjectSelection->getFirstRootObject(children_ok)->getScale();
-		rot = mObjectSelection->getFirstRootObject(children_ok)->getRotation();
+		pos = root_object->getPivotPositionAgent();
+		scale = root_object->getScale();
+		rot = root_object->getRotation();
 	}
 	else
 	{
-		pos = LLSelectMgr::getInstance()->getBBoxOfSelection().getCenterAgent();
-		scale = LLSelectMgr::getInstance()->getBBoxOfSelection().getExtentLocal();
-		rot = LLSelectMgr::getInstance()->getBBoxOfSelection().getRotation();
+		LLBBox bounding_box = LLSelectMgr::getInstance()->getBBoxOfSelection();
+		pos = bounding_box.getCenterAgent();
+		scale = bounding_box.getExtentLocal();
+		rot = bounding_box.getRotation();
 	}
-	if(sPivotPerc)
+	if (sPivotPerc)
 	{
 		LLVector3 add(
-			(-scale[VX]*0.5) + (scale[VX]*(sPivotX*0.01)),
-			(-scale[VY]*0.5) + (scale[VY]*(sPivotY*0.01)),
-			(-scale[VZ]*0.5) + (scale[VZ]*(sPivotZ*0.01)));
+			(-scale[VX] * 0.5) + (scale[VX] * (sPivotX * 0.01)),
+			(-scale[VY] * 0.5) + (scale[VY] * (sPivotY * 0.01)),
+			(-scale[VZ] * 0.5) + (scale[VZ] * (sPivotZ * 0.01)));
 		add = add * rot;
 		pos = pos + add;
 	}
@@ -386,6 +402,7 @@ LLVector3 LLManip::getPivotPoint()
 		pos = pos + add;
 	}
 	return pos;
+	// </FS:KC> Pivot point controls
 }
 
 
@@ -415,8 +432,10 @@ void LLManip::renderGuidelines(BOOL draw_x, BOOL draw_y, BOOL draw_z)
 		grid_rot.getAngleAxis(&angle_radians, &x, &y, &z);
 		gGL.rotatef(angle_radians * RAD_TO_DEG, x, y, z);
 
+		// <FS> OpenSim support
 		//F32 region_size = LLWorld::getInstance()->getRegionWidthInMeters();
 		F32 region_size = object->getRegion()->getWidth();
+		// </FS> OpenSim support
 
 		const F32 LINE_ALPHA = 0.33f;
 
