diff --git a/Source/indra/newview/llmanipscale.h b/Source/indra/newview/llmanipscale.h
index cfbb148..0850bec 100644
--- a/Source/indra/newview/llmanipscale.h
+++ b/Source/indra/newview/llmanipscale.h
@@ -92,7 +92,9 @@ private:
 	void			renderEdges( const LLBBox& local_bbox );
 	void			renderBoxHandle( F32 x, F32 y, F32 z );
 	void			renderAxisHandle( const LLVector3& start, const LLVector3& end );
-	void			renderGuidelinesPart( const LLBBox& local_bbox );
+// <FS:CR> FIRE-8882
+	//void			renderGuidelinesPart( const LLBBox& local_bbox );
+// </FS:CR>
 	void			renderSnapGuides( const LLBBox& local_bbox );
 
 	void			revert();
