diff --git a/Source/indra/newview/llmediadataclient.cpp b/Source/indra/newview/llmediadataclient.cpp
index 4ed0ffb..fce8c09 100644
--- a/Source/indra/newview/llmediadataclient.cpp
+++ b/Source/indra/newview/llmediadataclient.cpp
@@ -567,7 +567,7 @@ LLMediaDataClient::Responder::Responder(const request_ptr_t &request)
 }
 
 /*virtual*/
-void LLMediaDataClient::Responder::error(U32 status, const std::string& reason)
+void LLMediaDataClient::Responder::errorWithContent(U32 status, const std::string& reason, const LLSD& content)
 {
 	mRequest->stopTracking();
 
@@ -599,8 +599,8 @@ void LLMediaDataClient::Responder::error(U32 status, const std::string& reason)
 	}
 	else 
 	{
-		std::string msg = boost::lexical_cast<std::string>(status) + ": " + reason;
-		LL_WARNS("LLMediaDataClient") << *mRequest << " http error(" << msg << ")" << LL_ENDL;
+		LL_WARNS("LLMediaDataClient") << *mRequest << " http error [status:" 
+				<< status << "]:" << content << ")" << LL_ENDL;
 	}
 }
 
@@ -1023,7 +1023,7 @@ LLMediaDataClient::Responder *LLObjectMediaNavigateClient::RequestNavigate::crea
 }
 
 /*virtual*/
-void LLObjectMediaNavigateClient::Responder::error(U32 status, const std::string& reason)
+void LLObjectMediaNavigateClient::Responder::errorWithContent(U32 status, const std::string& reason, const LLSD& content)
 {
 	getRequest()->stopTracking();
 
@@ -1037,7 +1037,7 @@ void LLObjectMediaNavigateClient::Responder::error(U32 status, const std::string
 	// class
 	if (status == HTTP_SERVICE_UNAVAILABLE)
 	{
-		LLMediaDataClient::Responder::error(status, reason);
+		LLMediaDataClient::Responder::errorWithContent(status, reason, content);
 	}
 	else
 	{
