diff --git a/Source/indra/newview/llnamelistctrl.cpp b/Source/indra/newview/llnamelistctrl.cpp
index 8befdac..b0fbad3 100644
--- a/Source/indra/newview/llnamelistctrl.cpp
+++ b/Source/indra/newview/llnamelistctrl.cpp
@@ -68,7 +68,7 @@ LLNameListCtrl::LLNameListCtrl(const LLNameListCtrl::Params& p)
 {}
 
 // public
-void LLNameListCtrl::addNameItem(const LLUUID& agent_id, EAddPosition pos,
+LLScrollListItem* LLNameListCtrl::addNameItem(const LLUUID& agent_id, EAddPosition pos,
 								 BOOL enabled, const std::string& suffix)
 {
 	//llinfos << "LLNameListCtrl::addNameItem " << agent_id << llendl;
@@ -78,7 +78,7 @@ void LLNameListCtrl::addNameItem(const LLUUID& agent_id, EAddPosition pos,
 	item.enabled = enabled;
 	item.target = INDIVIDUAL;
 
-	addNameItemRow(item, pos, suffix);
+	return addNameItemRow(item, pos, suffix);
 }
 
 // virtual, public
@@ -204,7 +204,7 @@ BOOL LLNameListCtrl::handleToolTip(S32 x, S32 y, MASK mask)
 {
 	BOOL handled = FALSE;
 	S32 column_index = getColumnIndexFromOffset(x);
-	LLScrollListItem* hit_item = hitItem(x, y);
+	LLNameListItem* hit_item = dynamic_cast<LLNameListItem*>(hitItem(x, y));
 	if (hit_item
 		&& column_index == mNameColumnIndex)
 	{
@@ -228,7 +228,7 @@ BOOL LLNameListCtrl::handleToolTip(S32 x, S32 y, MASK mask)
 				LLCoordGL pos( sticky_rect.mRight - info_icon_size, sticky_rect.mTop - (sticky_rect.getHeight() - icon->getHeight())/2 );
 
 				// Should we show a group or an avatar inspector?
-				bool is_group = hit_item->getValue()["is_group"].asBoolean();
+				bool is_group = hit_item->isGroup();
 
 				LLToolTip::Params params;
 				params.background_visible( false );
@@ -271,10 +271,10 @@ void LLNameListCtrl::addGroupNameItem(LLNameListCtrl::NameItem& item, EAddPositi
 	addNameItemRow(item, pos);
 }
 
-void LLNameListCtrl::addNameItem(LLNameListCtrl::NameItem& item, EAddPosition pos)
+LLScrollListItem* LLNameListCtrl::addNameItem(LLNameListCtrl::NameItem& item, EAddPosition pos)
 {
 	item.target = INDIVIDUAL;
-	addNameItemRow(item, pos);
+	return addNameItemRow(item, pos);
 }
 
 LLScrollListItem* LLNameListCtrl::addElement(const LLSD& element, EAddPosition pos, void* userdata)
@@ -293,19 +293,12 @@ LLScrollListItem* LLNameListCtrl::addNameItemRow(
 	const std::string& suffix)
 {
 	LLUUID id = name_item.value().asUUID();
-	LLNameListItem* item = NULL;
-
-	// Store item type so that we can invoke the proper inspector.
-	// *TODO Vadim: Is there a more proper way of storing additional item data?
-	{
-		LLNameListCtrl::NameItem item_p(name_item);
-		item_p.value = LLSD().with("uuid", id).with("is_group", name_item.target() == GROUP);
-		item = new LLNameListItem(item_p);
-		LLScrollListCtrl::addRow(item, item_p, pos);
-	}
+	LLNameListItem* item = new LLNameListItem(name_item,name_item.target() == GROUP);
 
 	if (!item) return NULL;
 
+	LLScrollListCtrl::addRow(item, name_item, pos);
+
 	// use supplied name by default
 	std::string fullname = name_item.name;
 	switch(name_item.target)
@@ -334,12 +327,9 @@ LLScrollListItem* LLNameListCtrl::addNameItemRow(
 			else
 			{
 				// ...schedule a callback
-//				LLAvatarNameCache::get(id,
-//					boost::bind(&LLNameListCtrl::onAvatarNameCache,
-//						this, _1, _2));
-// [SL:KB] - Patch: UI-GroupPanel | Checked: 2011-05-30 (Catznip-2.6.0a) | Added: Catznip-2.6.0a
-				LLAvatarNameCache::get(id, boost::bind(&LLNameListItem::onAvatarNameCache, item, mNameColumnIndex, mShortNames, _1, _2));
-// [/SL:KB]
+				LLAvatarNameCache::get(id,
+					boost::bind(&LLNameListCtrl::onAvatarNameCache,
+						this, _1, _2, item->getHandle()));
 			}
 			break;
 		}
@@ -394,31 +384,29 @@ void LLNameListCtrl::removeNameItem(const LLUUID& agent_id)
 	}
 }
 
-//void LLNameListCtrl::onAvatarNameCache(const LLUUID& agent_id,
-//									   const LLAvatarName& av_name)
-//{
-//	std::string name;
-//	if (mShortNames)
-//		name = av_name.mDisplayName;
-//	else
-//		name = av_name.getCompleteName();
-//
-//	item_list::iterator iter;
-//	for (iter = getItemList().begin(); iter != getItemList().end(); iter++)
-//	{
-//		LLScrollListItem* item = *iter;
-//		if (item->getUUID() == agent_id)
-//		{
-//			LLScrollListCell* cell = item->getColumn(mNameColumnIndex);
-//			if (cell)
-//			{
-//				cell->setValue(name);
-//			}
-//		}
-//	}
-//
-//	dirtyColumns();
-//}
+void LLNameListCtrl::onAvatarNameCache(const LLUUID& agent_id,
+									   const LLAvatarName& av_name,
+									   LLHandle<LLNameListItem> item)
+{
+	std::string name;
+	if (mShortNames)
+		name = av_name.mDisplayName;
+	else
+		name = av_name.getCompleteName();
+
+	LLNameListItem* list_item = item.get();
+	if (list_item && list_item->getUUID() == agent_id)
+	{
+		LLScrollListCell* cell = list_item->getColumn(mNameColumnIndex);
+		if (cell)
+		{
+			cell->setValue(name);
+			setNeedsSort();
+		}
+	}
+	
+	dirtyColumns();
+}
 
 
 void LLNameListCtrl::updateColumns()
@@ -435,10 +423,7 @@ void LLNameListCtrl::updateColumns()
 	}
 }
 
-// [SL:KB] - Patch: UI-GroupPanel | Checked: 2011-05-30 (Catznip-2.6.0a) | Added: Catznip-2.6.0a
-void LLNameListItem::onAvatarNameCache(S32 idxNameColumn, bool useShortNames, const LLUUID& idAgent, const LLAvatarName& avName)
+void LLNameListCtrl::sortByName(BOOL ascending)
 {
-	if (LLScrollListCell* pCell = getColumn(idxNameColumn))
-		pCell->setValue( (useShortNames) ? avName.mDisplayName : avName.getCompleteName() );
+	sortByColumnIndex(mNameColumnIndex,ascending);
 }
-// [/SL:KB]
