diff --git a/Source/indra/newview/llnearbychathandler.h b/Source/indra/newview/llnearbychathandler.h
index 7a373cb..6d8b0c3 100644
--- a/Source/indra/newview/llnearbychathandler.h
+++ b/Source/indra/newview/llnearbychathandler.h
@@ -45,16 +45,10 @@ public:
 
 	virtual void processChat(const LLChat& chat_msg, const LLSD &args);
 
-	void updateFSUseNearbyChatConsole(const LLSD &data);
-
 protected:
 	virtual void onDeleteToast(LLToast* toast);
 	virtual void initChannel();
 
-	void onAvatarNameLookup(const LLUUID& agent_id, const LLAvatarName& av_name, const LLChat& chat_msg);
-
-	BOOL FSUseNearbyChatConsole;
-
 	static boost::scoped_ptr<LLEventPump> sChatWatcher;
 };
 
