diff --git a/Source/indra/newview/lloutfitslist.h b/Source/indra/newview/lloutfitslist.h
index a059873..2e3fb3f 100644
--- a/Source/indra/newview/lloutfitslist.h
+++ b/Source/indra/newview/lloutfitslist.h
@@ -108,6 +108,16 @@ public:
 	 */
 	bool hasItemSelected();
 
+	/**
+	Collapses all outfit accordions.
+	*/
+	void collapse_all_folders();
+	/**
+	Expands all outfit accordions.
+	*/
+	void expand_all_folders();
+
+
 private:
 
 	void onOutfitsRemovalConfirmation(const LLSD& notification, const LLSD& response);
