diff --git a/Source/indra/newview/llpanelvolumepulldown.cpp b/Source/indra/newview/llpanelvolumepulldown.cpp
index aea7b33..f3c5656 100644
--- a/Source/indra/newview/llpanelvolumepulldown.cpp
+++ b/Source/indra/newview/llpanelvolumepulldown.cpp
@@ -40,6 +40,9 @@
 #include "llfloaterpreference.h"
 #include "llsliderctrl.h"
 
+#include "llcheckboxctrl.h"
+#include "llviewercontrol.h"
+
 /* static */ const F32 LLPanelVolumePulldown::sAutoCloseFadeStartTimeSec = 4.0f;
 /* static */ const F32 LLPanelVolumePulldown::sAutoCloseTotalTimeSec = 5.0f;
 
@@ -54,13 +57,18 @@ LLPanelVolumePulldown::LLPanelVolumePulldown()
 
     mCommitCallbackRegistrar.add("Vol.setControlFalse", boost::bind(&LLPanelVolumePulldown::setControlFalse, this, _2));
 	mCommitCallbackRegistrar.add("Vol.GoAudioPrefs", boost::bind(&LLPanelVolumePulldown::onAdvancedButtonClick, this, _2));
+	// <FS:Ansariel> Missing callback function
+	mCommitCallbackRegistrar.add("Vol.SetSounds", boost::bind(&LLPanelVolumePulldown::setSounds, this));
+
 	buildFromFile( "panel_volume_pulldown.xml");
 }
 
 BOOL LLPanelVolumePulldown::postBuild()
 {
 	// set the initial volume-slider's position to reflect reality
-	LLSliderCtrl* volslider =  getChild<LLSliderCtrl>( "mastervolume" );
+	// <FS:Ansariel> Was renamed to "System Volume"
+	//LLSliderCtrl* volslider =  getChild<LLSliderCtrl>( "mastervolume" );
+	LLSliderCtrl* volslider =  getChild<LLSliderCtrl>( "System Volume" );
 	volslider->setValue(gSavedSettings.getF32("AudioLevelMaster"));
 
 	return LLPanel::postBuild();
@@ -146,3 +154,12 @@ void LLPanelVolumePulldown::draw()
 	}
 }
 
+// <FS:Ansariel> Missing callback function
+void LLPanelVolumePulldown::setSounds()
+{
+	// Disable Enable gesture/collisions sounds checkbox if the master sound is disabled
+	// or if sound effects are disabled.
+	getChild<LLCheckBoxCtrl>("gesture_audio_play_btn")->setEnabled(!gSavedSettings.getBOOL("MuteSounds"));
+	getChild<LLCheckBoxCtrl>("collisions_audio_play_btn")->setEnabled(!gSavedSettings.getBOOL("MuteSounds"));
+}
+// </FS:Ansariel> Missing callback function
