diff --git a/Source/indra/newview/llpanelvolumepulldown.h b/Source/indra/newview/llpanelvolumepulldown.h
index 0d86e6b..f5dcd94 100644
--- a/Source/indra/newview/llpanelvolumepulldown.h
+++ b/Source/indra/newview/llpanelvolumepulldown.h
@@ -48,6 +48,8 @@ class LLPanelVolumePulldown : public LLPanel
  private:
 	void setControlFalse(const LLSD& user_data);
 	void onAdvancedButtonClick(const LLSD& user_data);
+	// <FS:Ansariel> Missing callback function
+	void setSounds();
 
 	LLFrameTimer mHoverTimer;
 	static const F32 sAutoCloseFadeStartTimeSec;
