diff --git a/Source/indra/newview/llpanelcontents.cpp b/Source/indra/newview/llpanelcontents.cpp
index 3c7a228..ba59214 100644
--- a/Source/indra/newview/llpanelcontents.cpp
+++ b/Source/indra/newview/llpanelcontents.cpp
@@ -82,8 +82,6 @@ const char* LLPanelContents::PERMS_ANYONE_CONTROL_KEY = "perms_anyone_control";
 
 BOOL LLPanelContents::postBuild()
 {
-	LLRect rect = this->getRect();
-
 	setMouseOpaque(FALSE);
 
 	childSetAction("button new script",&LLPanelContents::onClickNewScript, this);
