diff --git a/Source/indra/newview/llpanelobject.cpp b/Source/indra/newview/llpanelobject.cpp
index 40c8196..73b9ca6 100644
--- a/Source/indra/newview/llpanelobject.cpp
+++ b/Source/indra/newview/llpanelobject.cpp
@@ -584,28 +584,29 @@ void LLPanelObject::getState( )
 	mBtnCopyParams->setEnabled( single_volume && objectp->permModify() );
 	mBtnPasteParams->setEnabled( single_volume && objectp->permModify() );
 
-	BOOL owners_identical;
 	LLUUID owner_id;
 	std::string owner_name;
-	owners_identical = LLSelectMgr::getInstance()->selectGetOwner(owner_id, owner_name);
+	LLSelectMgr::getInstance()->selectGetOwner(owner_id, owner_name);
 
 	// BUG? Check for all objects being editable?
 	S32 roots_selected = LLSelectMgr::getInstance()->getSelection()->getRootObjectCount();
 	BOOL editable = root_objectp->permModify();
 
 	// Select Single Message
-	getChildView("select_single")->setVisible( FALSE);
-	getChildView("edit_object")->setVisible( FALSE);
-	if (!editable || single_volume || selected_count <= 1)
-	{
-		getChildView("edit_object")->setVisible( TRUE);
-		getChildView("edit_object")->setEnabled(TRUE);
-	}
-	else
-	{
-		getChildView("select_single")->setVisible( TRUE);
-		getChildView("select_single")->setEnabled(TRUE);
-	}
+	// <FS:Ansariel> We don't have those in FS (2013-04-28)
+	//getChildView("select_single")->setVisible( FALSE);
+	//getChildView("edit_object")->setVisible( FALSE);
+	//if (!editable || single_volume || selected_count <= 1)
+	//{
+	//	getChildView("edit_object")->setVisible( TRUE);
+	//	getChildView("edit_object")->setEnabled(TRUE);
+	//}
+	//else
+	//{
+	//	getChildView("select_single")->setVisible( TRUE);
+	//	getChildView("select_single")->setEnabled(TRUE);
+	//}
+	// </FS:Ansariel>
 
 	BOOL is_flexible = volobjp && volobjp->isFlexible();
 	BOOL is_permanent = root_objectp->flagObjectPermanent();
@@ -2259,9 +2260,11 @@ void LLPanelObject::clearCtrls()
 	mLabelRadiusOffset->setEnabled( FALSE );
 	mLabelRevolutions->setEnabled( FALSE );
 
-	getChildView("select_single")->setVisible( FALSE);
-	getChildView("edit_object")->setVisible( TRUE);	
-	getChildView("edit_object")->setEnabled(FALSE);
+	// <FS:Ansariel> We don't have those in Firestorm (2013-04-28)
+	//getChildView("select_single")->setVisible( FALSE);
+	//getChildView("edit_object")->setVisible( TRUE);	
+	//getChildView("edit_object")->setEnabled(FALSE);
+	// </FS:Ansariel>
 	
 	getChildView("scale_hole")->setEnabled(FALSE);
 	getChildView("scale_taper")->setEnabled(FALSE);
