diff --git a/Source/indra/newview/llproductinforequest.cpp b/Source/indra/newview/llproductinforequest.cpp
index 93bf8b2..1390000 100644
--- a/Source/indra/newview/llproductinforequest.cpp
+++ b/Source/indra/newview/llproductinforequest.cpp
@@ -43,10 +43,10 @@ public:
 	}
 
 	//If we get back an error (not found, etc...), handle it here
-	virtual void error(U32 status, const std::string& reason)
+	virtual void errorWithContent(U32 status, const std::string& reason, const LLSD& content)
 	{
-		llwarns << "LLProductInfoRequest::error("
-		<< status << ": " << reason << ")" << llendl;
+		llwarns << "LLProductInfoRequest error [status:"
+				<< status << ":] " << content << llendl;
 	}
 };
 
