diff --git a/Source/indra/newview/llspeakers.cpp b/Source/indra/newview/llspeakers.cpp
index aa2f32f..fc8cff4 100644
--- a/Source/indra/newview/llspeakers.cpp
+++ b/Source/indra/newview/llspeakers.cpp
@@ -350,7 +350,7 @@ void LLSpeakerMgr::update(BOOL resort_ok)
 		LLUUID speaker_id = speaker_it->first;
 		LLSpeaker* speakerp = speaker_it->second;
 		
-		speaker_map_t::iterator  cur_speaker_it = speaker_it++;
+		speaker_it++;
 
 		if (voice_channel_active && LLVoiceClient::getInstance()->getVoiceEnabled(speaker_id))
 		{
@@ -726,9 +726,9 @@ public:
 		mSessionID = session_id;
 	}
 
-	virtual void error(U32 status, const std::string& reason)
+	virtual void errorWithContent(U32 status, const std::string& reason, const LLSD& content)
 	{
-		llwarns << status << ": " << reason << llendl;
+		llwarns << "ModerationResponder error [status:" << status << "]: " << content << llendl;
 
 		if ( gIMMgr )
 		{
