diff --git a/Source/indra/newview/llstartup.cpp b/Source/indra/newview/llstartup.cpp
index 8ba4805..fb682e8 100644
--- a/Source/indra/newview/llstartup.cpp
+++ b/Source/indra/newview/llstartup.cpp
@@ -2698,6 +2698,8 @@ void register_viewer_callbacks(LLMessageSystem* msg)
 	msg->setHandlerFuncFast(_PREHASH_CameraConstraint,		process_camera_constraint);
 	msg->setHandlerFuncFast(_PREHASH_AvatarSitResponse,		process_avatar_sit_response);
 	msg->setHandlerFunc("SetFollowCamProperties",			process_set_follow_cam_properties);
+	msg->setHandlerFunc("SetCameraProperties",				process_set_camera_properties);
+	msg->setHandlerFunc("SetFrustum",						process_set_frustum);
 	msg->setHandlerFunc("ClearFollowCamProperties",			process_clear_follow_cam_properties);
 
 	msg->setHandlerFuncFast(_PREHASH_ImprovedInstantMessage,	process_improved_im);
