diff --git a/Source/indra/newview/llstatusbar.h b/Source/indra/newview/llstatusbar.h
index 5d7bd47..db5b58c 100644
--- a/Source/indra/newview/llstatusbar.h
+++ b/Source/indra/newview/llstatusbar.h
@@ -148,8 +148,6 @@ private:
 	void onMouseEnterVolume();
 	void onMouseEnterNearbyMedia();
 	void onClickScreen(S32 x, S32 y);
-	void onModeChange(const LLSD& original_value, const LLSD& new_value);
-	void onModeChangeConfirm(const LLSD& original_value, const LLSD& new_value, const LLSD& notification, const LLSD& response);
 
 	static void onClickStreamToggle(void* data);		// ## Zi: Media/Stream separation
 	static void onClickMediaToggle(void* data);
@@ -203,6 +201,11 @@ private:
 	 */
 	void onParcelWLClicked();
 
+	/** <FS:CR> FIRE-5118 - Lightshare support
+	 * Handles clicks on the lightshare icon.
+	 */
+	void onLightshareClicked();
+	
 	/**
 	 * Called when agent changes the parcel.
 	 */
@@ -290,6 +293,7 @@ private:
 	void setParcelInfoText(const std::string& new_text);
 
 	void updateNetstatVisibility(const LLSD& data);
+	void updateVolumeControlsVisibility(const LLSD& data); // <FS:PP> Option to hide volume controls (sounds, media, stream) in upper right
 
 private:
 	LLTextBox	*mTextBalance;
@@ -307,9 +311,13 @@ private:
 	LLButton	*mStreamToggle;		// ## Zi: Media/Stream separation
 	LLButton	*mMediaToggle;
 	LLButton	*mBandwidthButton; // <FS:PP> FIRE-6287: Clicking on traffic indicator toggles Lag Meter window
-	LLView		*mScriptOut;
+	// <FS:Ansariel> Script debug
+	//LLView		*mScriptOut;
+	LLIconCtrl	*mScriptOut;
+	// </FS:Ansariel> Script debug
 	LLFrameTimer	mClockUpdateTimer;
 
+	S32				mVolumeIconsWidth; // <FS:PP> Option to hide volume controls (sounds, media, stream) in upper right
 	S32				mBalance;
 	S32				mHealth;
 	S32				mSquareMetersCredit;
@@ -328,6 +336,9 @@ private:
 	LLIconCtrl*				mParcelIcon[ICON_COUNT];
 	LLParcelChangeObserver*	mParcelChangedObserver;
 	LLButton* 				mPWLBtn;
+	// <FS:CR> FIRE-5118 - Lightshare support
+	LLButton*				mLightshareBtn;
+	// </FS:CR>
 	LLPanel*				mBalancePanel;
 	LLButton*				mBuyParcelBtn;
 	LLPanel*				mTimeMediaPanel;
@@ -343,7 +354,16 @@ private:
 	BOOL			mPathfindingFlashOn;
 	// </FS:Zi>
 
+	// <FS:Ansariel> Script debug
+	BOOL			mNearbyIcons;
+
 	bool	mRebakeStuck; // <FS:LO> FIRE-7639 - Stop the blinking after a while
+
+// <FS:Zi> Make hovering over parcel info actually work
+private:
+	void	onMouseEnterParcelInfo();
+	void	onMouseLeaveParcelInfo();
+// </FS:Zi>
 };
 
 // *HACK: Status bar owns your cached money balance. JC
