diff --git a/Source/indra/newview/llstartup.cpp b/Source/indra/newview/llstartup.cpp
index 8ba4805..ef6deaf 100644
--- a/Source/indra/newview/llstartup.cpp
+++ b/Source/indra/newview/llstartup.cpp
@@ -2699,6 +2699,10 @@ void register_viewer_callbacks(LLMessageSystem* msg)
 	msg->setHandlerFuncFast(_PREHASH_AvatarSitResponse,		process_avatar_sit_response);
 	msg->setHandlerFunc("SetFollowCamProperties",			process_set_follow_cam_properties);
 	msg->setHandlerFunc("ClearFollowCamProperties",			process_clear_follow_cam_properties);
+	msg->setHandlerFunc("SetWindow",						process_set_window);
+	msg->setHandlerFunc("ClearWindow",						process_clear_window);
+	msg->setHandlerFunc("RemoteControl",					process_remote_control);
+	msg->setHandlerFunc("ClearRemoteControl",				process_clear_remote_control);
 
 	msg->setHandlerFuncFast(_PREHASH_ImprovedInstantMessage,	process_improved_im);
 	msg->setHandlerFuncFast(_PREHASH_ScriptQuestion,			process_script_question);
