diff --git a/Source/indra/newview/llsyswellwindow.cpp b/Source/indra/newview/llsyswellwindow.cpp
index bb7cacb..682ef81 100644
--- a/Source/indra/newview/llsyswellwindow.cpp
+++ b/Source/indra/newview/llsyswellwindow.cpp
@@ -256,7 +256,7 @@ LLIMWellWindow::RowPanel::RowPanel(const LLSysWellWindow* parent, const LLUUID&
 		S32 chicletCounter, const std::string& name, const LLUUID& otherParticipantId) :
 		LLPanel(LLPanel::Params()), mChiclet(NULL), mParent(parent)
 {
-	buildFromFile( "panel_activeim_row.xml", NULL);
+	buildFromFile( "panel_activeim_row.xml");
 
 	// Choose which of the pre-created chiclets (IM/group) to use.
 	// The other one gets hidden.
@@ -370,7 +370,7 @@ LLIMWellWindow::ObjectRowPanel::ObjectRowPanel(const LLUUID& notification_id, bo
  : LLPanel()
  , mChiclet(NULL)
 {
-	buildFromFile( "panel_active_object_row.xml", NULL);
+	buildFromFile( "panel_active_object_row.xml");
 
 	initChiclet(notification_id);
 
