diff --git a/Source/indra/newview/lltexturectrl.h b/Source/indra/newview/lltexturectrl.h
index c4b18a5..f294658 100644
--- a/Source/indra/newview/lltexturectrl.h
+++ b/Source/indra/newview/lltexturectrl.h
@@ -186,6 +186,9 @@ public:
 
 	LLViewerFetchedTexture* getTexture() { return mTexturep; }
 
+	// <FS:Ansariel> Mask texture if desired
+	void setIsMasked(BOOL masked) { mIsMasked = masked; }
+
 private:
 	BOOL allowDrop(LLInventoryItem* item);
 	BOOL doDrop(LLInventoryItem* item);
@@ -219,7 +222,12 @@ private:
 	BOOL					 	mShowLoadingPlaceholder;
 	std::string				 	mLoadingPlaceholderString;
 	S32						 	mLabelWidth;
-	BOOL						mPreviewMode; // If true open texture in preview floater
+
+	// <FS:Ansariel> Texture preview mode
+	BOOL						mPreviewMode;
+
+	// <FS:Ansariel> Mask texture if desired
+	BOOL						mIsMasked;
 };
 
 // XUI HACK: When floaters converted, switch this file to lltexturepicker.h/cpp
