diff --git a/Source/indra/newview/lltoolpie.h b/Source/indra/newview/lltoolpie.h
index 213fc98..0bd8500 100644
--- a/Source/indra/newview/lltoolpie.h
+++ b/Source/indra/newview/lltoolpie.h
@@ -96,6 +96,13 @@ private:
 	BOOL handleTooltipLand(std::string line, std::string tooltip_msg);
 	BOOL handleTooltipObject( LLViewerObject* hover_object, std::string line, std::string tooltip_msg);
 
+	// <FS:ND> FIRE-10276; handleTooltipObject can be called during name resolution (LLAvatarNameCache), then hover_object can lon gbe destroyed and the pointer invalid.
+	// To circumvent this just pass the id and try to fetch the object from gObjectList.
+
+	BOOL handleTooltipObjectById( LLUUID hoverObjectId, std::string line, std::string tooltip_msg);
+
+	// </FS:ND>
+
 	void steerCameraWithMouse(S32 x, S32 y);
 	void startCameraSteering();
 	void stopCameraSteering();
@@ -121,6 +128,15 @@ private:
 	LLSafeHandle<LLObjectSelection> mLeftClickSelection;
 	BOOL				mClickActionBuyEnabled;
 	BOOL				mClickActionPayEnabled;
+
+	// <FS:ND> Keep track of name resolutions we made and delete them if needed to avoid crashing if this instance dies.
+private:
+	typedef boost::signals2::connection tNamecacheConnection;
+	std::vector< tNamecacheConnection > mNamecacheConnections;
+
+public:
+	~LLToolPie();
+	// </FS:ND>
 };
 
 #endif
