diff --git a/Source/indra/newview/lltoastnotifypanel.cpp b/Source/indra/newview/lltoastnotifypanel.cpp
index 5a0197c..417dc24 100644
--- a/Source/indra/newview/lltoastnotifypanel.cpp
+++ b/Source/indra/newview/lltoastnotifypanel.cpp
@@ -40,7 +40,10 @@
 #include "lltrans.h"
 #include "llnotificationsutil.h"
 #include "llviewermessage.h"
-#include "llimfloater.h"
+// <FS:Ansariel> [FS communication UI]
+//#include "llimfloater.h"
+#include "fsfloaterim.h"
+// </FS:Ansariel> [FS communication UI]
 
 const S32 BOTTOM_PAD = VPAD * 3;
 const S32 IGNORE_BTN_TOP_DELTA = 3*VPAD;//additional ignore_btn padding
@@ -309,12 +312,12 @@ void LLToastNotifyPanel::updateButtonsLayout(const std::vector<index_button_pair
 	S32 bottom_offset = mIsScriptDialog ? (BTN_HEIGHT + IGNORE_BTN_TOP_DELTA + BOTTOM_PAD) : BOTTOM_PAD;
 	S32 max_width = mControlPanel->getRect().getWidth();
 	LLButton* ignore_btn = NULL;
-	// LLButton* mute_btn = NULL; *spam protection is part of FS now. May be a debug setting in future*
+	LLButton* mute_btn = NULL;
 	for (std::vector<index_button_pair_t>::const_iterator it = buttons.begin(); it != buttons.end(); it++)
 	{
 		if (-2 == it->first)
 		{
-	//		mute_btn = it->second; *Removing Mute Button*
+			mute_btn = it->second;
 			continue;
 		}
 		if (it->first == -1)
@@ -338,36 +341,30 @@ void LLToastNotifyPanel::updateButtonsLayout(const std::vector<index_button_pair
 	}
 
 	U32 ignore_btn_width = 0;
+	U32 mute_btn_pad = 0;
 	if (mIsScriptDialog && ignore_btn != NULL)
 	{
 		LLRect ignore_btn_rect(ignore_btn->getRect());
-		S32 buttons_per_row = max_width / BUTTON_WIDTH; //assume that h_pad far less than BUTTON_WIDTH
-		S32 ignore_btn_left = buttons_per_row * BUTTON_WIDTH + (buttons_per_row	- 1) * h_pad - ignore_btn_rect.getWidth();
-		if (ignore_btn_left + ignore_btn_rect.getWidth() > max_width)// make sure that the ignore button is in panel
-		{
-			ignore_btn_left = max_width - ignore_btn_rect.getWidth() - 2 * HPAD;
-		}
+		S32 ignore_btn_left = max_width - ignore_btn_rect.getWidth();
 		ignore_btn_rect.setOriginAndSize(ignore_btn_left, BOTTOM_PAD,// always move ignore button at the bottom
 				ignore_btn_rect.getWidth(), ignore_btn_rect.getHeight());
 		ignore_btn->setRect(ignore_btn_rect);
 		ignore_btn_width = ignore_btn_rect.getWidth();
 		mControlPanel->addChild(ignore_btn, -1);
+		mute_btn_pad = 4 * HPAD; //only use a 4 * HPAD padding if an ignore button exists
 	}
-	// Commenting all out as mute button is disabled
-	// if (mIsScriptDialog && mute_btn != NULL)
+	// FIRE-3948: Commenting all out as mute button is disabled (FS:MS)
+	// <FS:Ansariel> Undo the removal and make it optional after I was looking for the mute button on spammy dialogs!
+	//if (mIsScriptDialog && mute_btn != NULL)
+	if (mIsScriptDialog && mute_btn != NULL && !gSavedSettings.getBOOL("FSRemoveScriptBlockButton"))
 	{
-	//	LLRect mute_btn_rect(mute_btn->getRect());
-	//	S32 buttons_per_row = max_width / BUTTON_WIDTH; //assume that h_pad far less than BUTTON_WIDTH
+		LLRect mute_btn_rect(mute_btn->getRect());
 		// Place mute (Block) button to the left of the ignore button.
-	//	S32 mute_btn_left = buttons_per_row * BUTTON_WIDTH + (buttons_per_row	- 1) * h_pad - mute_btn_rect.getWidth() - ignore_btn_width - (h_pad / 2);
-	//	if (mute_btn_left + mute_btn_rect.getWidth() > max_width) // make sure that the mute button is in panel
-	//	{
-	//		mute_btn_left = max_width - mute_btn_rect.getWidth() - 2 * HPAD;
-	//	}
-	//	mute_btn_rect.setOriginAndSize(mute_btn_left, BOTTOM_PAD,// always move mute button at the bottom
-	//			mute_btn_rect.getWidth(), mute_btn_rect.getHeight());
-	//	mute_btn->setRect(mute_btn_rect);
-	//	mControlPanel->addChild(mute_btn); 
+		S32 mute_btn_left = max_width - mute_btn_rect.getWidth() - ignore_btn_width - mute_btn_pad;
+		mute_btn_rect.setOriginAndSize(mute_btn_left, BOTTOM_PAD,// always move mute button at the bottom
+				mute_btn_rect.getWidth(), mute_btn_rect.getHeight());
+		mute_btn->setRect(mute_btn_rect);
+		mControlPanel->addChild(mute_btn); 
 	}
 }
 void LLToastNotifyPanel::adjustPanelForScriptNotice(S32 button_panel_width, S32 button_panel_height)
@@ -380,8 +377,6 @@ void LLToastNotifyPanel::adjustPanelForScriptNotice(S32 button_panel_width, S32
 
 void LLToastNotifyPanel::adjustPanelForTipNotice()
 {
-	LLRect info_rect = mInfoPanel->getRect();
-	LLRect this_rect = getRect();
 	//we don't need display ControlPanel for tips because they doesn't contain any buttons. 
 	mControlPanel->setVisible(FALSE);
 	reshape(getRect().getWidth(), mInfoPanel->getRect().getHeight());
@@ -571,7 +566,10 @@ LLIMToastNotifyPanel::~LLIMToastNotifyPanel()
 	// This may happened when IM floater reloads messages, exactly when user
 	// changes layout of IM chat log(disable/enable plaintext mode).
 	// See EXT-6500
-	LLIMFloater* im_floater = LLIMFloater::findInstance(mSessionID);
+	// <FS:Ansariel> [FS communication UI]
+	//LLIMFloater* im_floater = LLIMFloater::findInstance(mSessionID);
+	FSFloaterIM* im_floater = FSFloaterIM::findInstance(mSessionID);
+	// </FS:Ansariel> [FS communication UI]
 	if (im_floater != NULL && !im_floater->isDead())
 	{
 		mCloseNotificationOnDestroy = false;
