diff --git a/Source/indra/newview/llviewergenericmessage.cpp b/Source/indra/newview/llviewergenericmessage.cpp
index f8a2be1..d804831 100644
--- a/Source/indra/newview/llviewergenericmessage.cpp
+++ b/Source/indra/newview/llviewergenericmessage.cpp
@@ -34,6 +34,7 @@
 #include "message.h"
 
 #include "llagent.h"
+#include "fslightshare.h"	// <FS:CR> FIRE-5118 - Lightshare support
 
 
 LLDispatcher gGenericDispatcher;
@@ -74,6 +75,14 @@ void send_generic_message(const std::string& method,
 
 void process_generic_message(LLMessageSystem* msg, void**)
 {
+// <FS:CR> FIRE-5118 - Lightshare support
+	std::string method;
+	msg->getStringFast(_PREHASH_MethodData, _PREHASH_Method, method);
+	if (method == "Windlight")
+		FSLightshare::getInstance()->processLightshareMessage(msg);
+	else if (method == "WindlightRefresh")
+		FSLightshare::getInstance()->processLightshareRefresh();
+// </FS:CR> Lightshare support
 	LLUUID agent_id;
 	msg->getUUID("AgentData", "AgentID", agent_id);
 	if (agent_id != gAgent.getID())
