diff --git a/Source/indra/newview/llviewershadermgr.h b/Source/indra/newview/llviewershadermgr.h
index d6dd645..bd3b5e5 100644
--- a/Source/indra/newview/llviewershadermgr.h
+++ b/Source/indra/newview/llviewershadermgr.h
@@ -55,7 +55,7 @@ public:
 	BOOL loadShadersInterface();
 	BOOL loadShadersWindLight();
 	BOOL loadTransformShaders();
-
+	
 	std::vector<S32> mVertexShaderLevel;
 	S32	mMaxAvatarShaderLevel;
 
@@ -229,7 +229,6 @@ extern LLGLSLShader			gSplatTextureRectProgram;
 extern LLGLSLShader			gGlowCombineFXAAProgram;
 extern LLGLSLShader			gDebugProgram;
 extern LLGLSLShader			gClipProgram;
-extern LLGLSLShader			gAlphaMaskProgram;
 
 //output tex0[tc0] + tex1[tc1]
 extern LLGLSLShader			gTwoTextureAddProgram;
@@ -361,5 +360,6 @@ extern LLGLSLShader			gDeferredWLSkyProgram;
 extern LLGLSLShader			gDeferredWLCloudProgram;
 extern LLGLSLShader			gDeferredStarProgram;
 extern LLGLSLShader			gNormalMapGenProgram;
+extern LLGLSLShader         gVignettePost;	// <FS:CR> Import Vignette from Exodus
 
 #endif
