diff --git a/Source/indra/newview/llviewermessage.h b/Source/indra/newview/llviewermessage.h
index ae0d5a0..64d794e 100644
--- a/Source/indra/newview/llviewermessage.h
+++ b/Source/indra/newview/llviewermessage.h
@@ -107,6 +107,8 @@ void process_camera_constraint(LLMessageSystem *mesgsys, void **user_data);
 void process_avatar_sit_response(LLMessageSystem *mesgsys, void **user_data);
 void process_set_follow_cam_properties(LLMessageSystem *mesgsys, void **user_data);
 void process_clear_follow_cam_properties(LLMessageSystem *mesgsys, void **user_data);
+void process_set_window(LLMessageSystem *mesgsys, void **user_data);
+void process_clear_window(LLMessageSystem *mesgsys, void **user_data);
 void process_name_value(LLMessageSystem *mesgsys, void **user_data);
 void process_remove_name_value(LLMessageSystem *mesgsys, void **user_data);
 void process_kick_user(LLMessageSystem *msg, void** /*user_data*/);
