diff --git a/Source/indra/newview/llviewermedia.cpp b/Source/indra/newview/llviewermedia.cpp
index 787f01f..a1e92bb 100644
--- a/Source/indra/newview/llviewermedia.cpp
+++ b/Source/indra/newview/llviewermedia.cpp
@@ -316,9 +316,13 @@ public:
 	/* virtual */ void completedHeader(U32 status, const std::string& reason, const LLSD& content)
 	{
 		LL_WARNS("MediaAuth") << "status = " << status << ", reason = " << reason << LL_ENDL;
-		LL_WARNS("MediaAuth") << content << LL_ENDL;
+
+		LLSD stripped_content = content;
+		stripped_content.erase("set-cookie");
+		LL_WARNS("MediaAuth") << stripped_content << LL_ENDL;
 
 		std::string cookie = content["set-cookie"].asString();
+		LL_DEBUGS("MediaAuth") << "cookie = " << cookie << LL_ENDL;
 
 		LLViewerMedia::getCookieStore()->setCookiesFromHost(cookie, mHost);
 
@@ -1188,12 +1192,9 @@ void LLViewerMedia::clearAllCookies()
 	LLDirIterator dir_iter(base_dir, "*_*");
 	while (dir_iter.next(filename))
 	{
-		target = base_dir;
-		target += filename;
-		target += gDirUtilp->getDirDelimiter();
-		target += "browser_profile";
-		target += gDirUtilp->getDirDelimiter();
-		target += "cookies";
+		target = gDirUtilp->add(base_dir, filename);
+		gDirUtilp->append(target, "browser_profile");
+		gDirUtilp->append(target, "cookies");
 		lldebugs << "target = " << target << llendl;
 		if(LLFile::isfile(target))
 		{	
@@ -1201,10 +1202,8 @@ void LLViewerMedia::clearAllCookies()
 		}
 		
 		// Other accounts may have new-style cookie files too -- delete them as well
-		target = base_dir;
-		target += filename;
-		target += gDirUtilp->getDirDelimiter();
-		target += PLUGIN_COOKIE_FILE_NAME;
+		target = gDirUtilp->add(base_dir, filename);
+		gDirUtilp->append(target, PLUGIN_COOKIE_FILE_NAME);
 		lldebugs << "target = " << target << llendl;
 		if(LLFile::isfile(target))
 		{	
