diff --git a/Source/indra/newview/llviewerstats.h b/Source/indra/newview/llviewerstats.h
index 554e4d6..e74fb36 100644
--- a/Source/indra/newview/llviewerstats.h
+++ b/Source/indra/newview/llviewerstats.h
@@ -33,92 +33,90 @@
 class LLViewerStats : public LLSingleton<LLViewerStats>
 {
 public:
-	LLStat mKBitStat;
-	LLStat mLayersKBitStat;
-	LLStat mObjectKBitStat;
-	LLStat mAssetKBitStat;
-	LLStat mTextureKBitStat;
-	LLStat mVFSPendingOperations;
-	LLStat mObjectsDrawnStat;
-	LLStat mObjectsCulledStat;
-	LLStat mObjectsTestedStat;
-	LLStat mObjectsComparedStat;
-	LLStat mObjectsOccludedStat;
-	LLStat mFPSStat;
-	LLStat mPacketsInStat;
-	LLStat mPacketsLostStat;
-	LLStat mPacketsOutStat;
-	LLStat mPacketsLostPercentStat;
-	LLStat mTexturePacketsStat;
-	LLStat mActualInKBitStat;	// From the packet ring (when faking a bad connection)
-	LLStat mActualOutKBitStat;	// From the packet ring (when faking a bad connection)
-	LLStat mTrianglesDrawnStat;
+	LLStat	mKBitStat,
+			mLayersKBitStat,
+			mObjectKBitStat,
+			mAssetKBitStat,
+			mTextureKBitStat,
+			mVFSPendingOperations,
+			mObjectsDrawnStat,
+			mObjectsCulledStat,
+			mObjectsTestedStat,
+			mObjectsComparedStat,
+			mObjectsOccludedStat,
+			mFPSStat,
+			mPacketsInStat,
+			mPacketsLostStat,
+			mPacketsOutStat,
+			mPacketsLostPercentStat,
+			mTexturePacketsStat,
+			mActualInKBitStat,	// From the packet ring (when faking a bad connection)
+			mActualOutKBitStat,	// From the packet ring (when faking a bad connection)
+			mTrianglesDrawnStat;
 
 	// Simulator stats
-	LLStat mSimTimeDilation;
-
-	LLStat mSimFPS;
-	LLStat mSimPhysicsFPS;
-	LLStat mSimAgentUPS;
-	LLStat mSimScriptEPS;
-
-	LLStat mSimFrameMsec;
-	LLStat mSimNetMsec;
-	LLStat mSimSimOtherMsec;
-	LLStat mSimSimPhysicsMsec;
-
-	LLStat mSimSimPhysicsStepMsec;
-	LLStat mSimSimPhysicsShapeUpdateMsec;
-	LLStat mSimSimPhysicsOtherMsec;
-
-	LLStat mSimSimAIStepMsec;
-	LLStat mSimSimSkippedSilhouetteSteps;
-	LLStat mSimSimPctSteppedCharacters;
-
-	LLStat mSimAgentMsec;
-	LLStat mSimImagesMsec;
-	LLStat mSimScriptMsec;
-	LLStat mSimSpareMsec;
-	LLStat mSimSleepMsec;
-	LLStat mSimPumpIOMsec;
-
-	LLStat mSimMainAgents;
-	LLStat mSimChildAgents;
-	LLStat mSimObjects;
-	LLStat mSimActiveObjects;
-	LLStat mSimActiveScripts;
-	LLStat mSimPctScriptsRun;
-
-	LLStat mSimInPPS;
-	LLStat mSimOutPPS;
-	LLStat mSimPendingDownloads;
-	LLStat mSimPendingUploads;
-	LLStat mSimPendingLocalUploads;
-	LLStat mSimTotalUnackedBytes;
-
-	LLStat mPhysicsPinnedTasks;
-	LLStat mPhysicsLODTasks;
-	LLStat mPhysicsMemoryAllocated;
-
-	LLStat mSimPingStat;
-
-	LLStat mNumImagesStat;
-	LLStat mNumRawImagesStat;
-	LLStat mGLTexMemStat;
-	LLStat mGLBoundMemStat;
-	LLStat mRawMemStat;
-	LLStat mFormattedMemStat;
-
-	LLStat mNumObjectsStat;
-	LLStat mNumActiveObjectsStat;
-	LLStat mNumNewObjectsStat;
-	LLStat mNumSizeCulledStat;
-	LLStat mNumVisCulledStat;
+	LLStat	mSimTimeDilation,
+
+			mSimFPS,
+			mSimPhysicsFPS,
+			mSimAgentUPS,
+			mSimScriptEPS,
+
+			mSimFrameMsec,
+			mSimNetMsec,
+			mSimSimOtherMsec,
+			mSimSimPhysicsMsec,
+
+			mSimSimPhysicsStepMsec,
+			mSimSimPhysicsShapeUpdateMsec,
+			mSimSimPhysicsOtherMsec,
+			mSimSimAIStepMsec,
+			mSimSimSkippedSilhouetteSteps,
+			mSimSimPctSteppedCharacters,
+
+			mSimAgentMsec,
+			mSimImagesMsec,
+			mSimScriptMsec,
+			mSimSpareMsec,
+			mSimSleepMsec,
+			mSimPumpIOMsec,
+
+			mSimMainAgents,
+			mSimChildAgents,
+			mSimObjects,
+			mSimActiveObjects,
+			mSimActiveScripts,
+			mSimPctScriptsRun,
+
+			mSimInPPS,
+			mSimOutPPS,
+			mSimPendingDownloads,
+			mSimPendingUploads,
+			mSimPendingLocalUploads,
+			mSimTotalUnackedBytes,
+
+			mPhysicsPinnedTasks,
+			mPhysicsLODTasks,
+			mPhysicsMemoryAllocated,
+
+			mSimPingStat,
+
+			mNumImagesStat,
+			mNumRawImagesStat,
+			mGLTexMemStat,
+			mGLBoundMemStat,
+			mRawMemStat,
+			mFormattedMemStat,
+
+			mNumObjectsStat,
+			mNumActiveObjectsStat,
+			mNumNewObjectsStat,
+			mNumSizeCulledStat,
+			mNumVisCulledStat;
 
 	void resetStats();
 public:
-	// If you change this, please also add a corresponding text label
-	// in statTypeToText in llviewerstats.cpp
+	// If you change this, please also add a corresponding text label in llviewerstats.cpp
 	enum EStatType
 	{
 		ST_VERSION = 0,
@@ -183,7 +181,6 @@ public:
 		ST_COUNT = 58
 	};
 
-
 	LLViewerStats();
 	~LLViewerStats();
 
@@ -282,23 +279,22 @@ public:
 
 	// Phase tracking (originally put in for avatar rezzing), tracking
 	// progress of active/completed phases for activities like outfit changing.
-	typedef std::map<std::string,LLFrameTimer>	phase_map_t;
+	typedef std::map<std::string,LLTimer>	phase_map_t;
 	typedef std::map<std::string,StatsAccumulator>	phase_stats_t;
 	class PhaseMap
 	{
 	private:
 		phase_map_t mPhaseMap;
-		static phase_stats_t sStats;
 	public:
 		PhaseMap();
-		LLFrameTimer& 	getPhaseTimer(const std::string& phase_name);
+		LLTimer&	 	getPhaseTimer(const std::string& phase_name);
+		bool 			getPhaseValues(const std::string& phase_name, F32& elapsed, bool& completed);
 		void			startPhase(const std::string& phase_name);
 		void			stopPhase(const std::string& phase_name);
-		void			stopAllPhases();
 		void			clearPhases();
 		LLSD			dumpPhases();
-		static StatsAccumulator& getPhaseStats(const std::string& phase_name);
-		static void recordPhaseStat(const std::string& phase_name, F32 value);
+		phase_map_t::iterator begin() { return mPhaseMap.begin(); }
+		phase_map_t::iterator end() { return mPhaseMap.end(); }
 	};
 
 private:
@@ -310,14 +306,10 @@ private:
 static const F32 SEND_STATS_PERIOD = 300.0f;
 
 // The following are from (older?) statistics code found in appviewer.
-void init_statistics();
-void reset_statistics();
-void output_statistics(void*);
-void update_statistics(U32 frame_count);
+void update_statistics();
 void send_stats();
 
-extern std::map<S32,LLFrameTimer> gDebugTimers;
-extern std::map<S32,std::string> gDebugTimerLabel;
+extern LLFrameTimer gTextureTimer;
 extern U32	gTotalTextureBytes;
 extern U32  gTotalObjectBytes;
 extern U32  gTotalTextureBytesPerBoostLevel[] ;
