diff --git a/Source/indra/newview/llviewerprecompiledheaders.h b/Source/indra/newview/llviewerprecompiledheaders.h
index 6c8a827..0316f79 100644
--- a/Source/indra/newview/llviewerprecompiledheaders.h
+++ b/Source/indra/newview/llviewerprecompiledheaders.h
@@ -59,8 +59,6 @@
 #include "indra_constants.h"
 #include "llinitparam.h"
 
-//#include "linden_common.h"
-//#include "llpreprocessor.h"
 #include "llallocator.h"
 #include "llapp.h"
 #include "llcriticaldamp.h"
@@ -77,10 +75,8 @@
 #include "llprocessor.h"
 #include "llrefcount.h"
 #include "llsafehandle.h"
-//#include "llsecondlifeurls.h"
 #include "llsd.h"
 #include "llsingleton.h"
-#include "llstat.h"
 #include "llstl.h"
 #include "llstrider.h"
 #include "llstring.h"
@@ -88,11 +84,8 @@
 #include "llthread.h"
 #include "lltimer.h"
 #include "lluuidhashmap.h"
-//#include "processor.h"
 #include "stdenums.h"
 #include "stdtypes.h"
-//#include "string_table.h"
-//#include "timer.h"
 #include "timing.h"
 #include "u64.h"
 
