diff --git a/Source/llviewermessage.cpp b/Source/llviewermessage.cpp
index 0f4fa06..0c399e8 100644
--- a/Source/llviewermessage.cpp
+++ b/Source/llviewermessage.cpp
@@ -6145,6 +6145,65 @@ void process_set_follow_cam_properties(LLMessageSystem *mesgsys, void **user_dat
 }
 //end Ventrella 
 
+void process_set_camera_properties(LLMessageSystem *msg, void ** )
+{
+	F32 frustumOffsetH;
+	F32 frustumOffsetV;
+	F32 frustumNear;
+	F32 viewAngle;
+	F32 aspectRatio; //width / height
+	BOOL aspectSet;
+	BOOL setNear;
+
+	msg->getF32( "CameraProperty", "FrustumOffsetH", frustumOffsetH );
+	msg->getF32( "CameraProperty", "FrustumOffsetV", frustumOffsetV);
+	msg->getF32( "CameraProperty", "FrustumNear", frustumNear);
+	msg->getF32( "CameraProperty", "CameraAngle", viewAngle );
+	msg->getF32( "CameraProperty", "AspectRatio", aspectRatio );
+	msg->getBOOL( "CameraProperty", "AspectSet", aspectSet );
+	msg->getBOOL( "CameraProperty", "SetNear", setNear);
+
+	gSavedSettings.setF32("FrustumOffsetH", frustumOffsetH);
+	gSavedSettings.setF32("FrustumOffsetV", frustumOffsetV);
+	gSavedSettings.setF32("FrustumNear", frustumNear);
+	gSavedSettings.setF32("CameraAngle", viewAngle);
+	gSavedSettings.setBOOL("ControlNear", setNear);
+
+	if (aspectSet)
+		LLViewerCamera::getInstance()->setAspect(aspectRatio);
+}
+
+void process_set_frustum(LLMessageSystem *msg, void ** )
+{
+	BOOL controlFrustum;
+
+	msg->getBOOL( "Frustum", "ControlFrustum", controlFrustum);
+	gSavedSettings.setBOOL("ControlFrustum", controlFrustum);
+
+	if (!controlFrustum)
+		return;
+
+	F32 x1;
+	F32 x2;
+	F32 y1;
+	F32 y2;
+	F32 dn;
+	F32 df;
+
+	msg->getF32( "Frustum", "x1", x1);
+	msg->getF32( "Frustum", "x2", x2);
+	msg->getF32( "Frustum", "y1", y1);
+	msg->getF32( "Frustum", "y2", y2);
+	msg->getF32( "Frustum", "dn", dn);
+	msg->getF32( "Frustum", "df", df);
+
+	gSavedSettings.setF32("x1", x1);
+	gSavedSettings.setF32("x2", x2);
+	gSavedSettings.setF32("y1", y1);
+	gSavedSettings.setF32("y2", y2);
+	gSavedSettings.setF32("dn", dn);
+	gSavedSettings.setF32("df", df);
+}
 
 // Culled from newsim lltask.cpp
 void process_name_value(LLMessageSystem *mesgsys, void **user_data)
