diff --git a/Source/indra/newview/llviewerkeyboard.cpp b/Source/indra/newview/llviewerkeyboard.cpp
index 5632719..e153a62 100644
--- a/Source/indra/newview/llviewerkeyboard.cpp
+++ b/Source/indra/newview/llviewerkeyboard.cpp
@@ -33,7 +33,7 @@
 #include "llagentcamera.h"
 // <FS:Zi> Remove floating chat bar
 // #include "llnearbychatbar.h"
-#include "llnearbychathub.h"
+#include "fsnearbychathub.h"
 #include "lllineeditor.h"
 // </FS:Zi>
 #include "llviewercontrol.h"
@@ -78,7 +78,7 @@ void agent_jump( EKeystate s )
 	// <FS:Ansariel> Chalice Yao's crouch toggle
 	if (gSavedSettings.getBOOL("FSCrouchToggleStatus"))
 	{
-		gSavedSettings.setBOOL("FSCrouchToggleStatus", false);
+		gSavedSettings.setBOOL("FSCrouchToggleStatus", FALSE);
 	}
 	// </FS:Ansariel>
 
@@ -105,11 +105,11 @@ void agent_push_down( EKeystate s )
 	{
 		if (gSavedSettings.getBOOL("FSCrouchToggleStatus"))
 		{
-			gSavedSettings.setBOOL("FSCrouchToggleStatus", false);
+			gSavedSettings.setBOOL("FSCrouchToggleStatus", FALSE);
 		}
 		else
 		{
-			gSavedSettings.setBOOL("FSCrouchToggleStatus", true);
+			gSavedSettings.setBOOL("FSCrouchToggleStatus", TRUE);
 			gAgent.moveUp(-1);
 		}
 	}
@@ -299,12 +299,6 @@ void agent_toggle_fly( EKeystate s )
 	if (KEYSTATE_DOWN == s )
 	{
 		LLAgent::toggleFlying();
-		// <FS:Ansariel> Chalice Yao's crouch toggle
-		if (gSavedSettings.getBOOL("FSCrouchToggleStatus"))
-		{
-			gSavedSettings.setBOOL("FSCrouchToggleStatus", false);
-		}
-		// </FS:Ansariel>
 	}
 }
 
@@ -582,7 +576,7 @@ void start_chat( EKeystate s )
 	// start chat
 	// <FS:Zi> Remove floating chat bar
 	// LLNearbyChatBar::startChat(NULL);
-	LLNearbyChat::instance().showDefaultChatBar(TRUE);
+	FSNearbyChat::instance().showDefaultChatBar(TRUE);
 	// </FS:Zi>
 }
 
@@ -620,7 +614,7 @@ void start_gesture( EKeystate s )
  		//	LLNearbyChat::startChat(NULL);
  		//}
 
-		LLNearbyChat::instance().showDefaultChatBar(TRUE,"/");
+		FSNearbyChat::instance().showDefaultChatBar(TRUE,"/");
  		// </FS:Ansariel>
 	}
 }
