diff --git a/Source/indra/newview/llviewercontrol.cpp b/Source/indra/newview/llviewercontrol.cpp
index e93d30a..825588b 100644
--- a/Source/indra/newview/llviewercontrol.cpp
+++ b/Source/indra/newview/llviewercontrol.cpp
@@ -71,16 +71,23 @@
 #include "llpaneloutfitsinventory.h"
 #include "llpanellogin.h"
 #include "llpaneltopinfobar.h"
-#include "llcombobox.h"
 #include "llspellcheck.h"
-#include "llstatusbar.h"
+#include "llslurl.h"
+#include "llstartup.h"
 #include "llupdaterservice.h"
-#include "llnotificationsutil.h"
 
 // NaCl - Antispam Registry
 #include "NACLantispam.h"
 // NaCl End
+//#include "llcombobox.h"
 #include "llnetmap.h"
+#include "llnotificationsutil.h"
+#include "llstatusbar.h"
+#include "llfloaterreg.h"
+#include "llfloatersidepanelcontainer.h"
+#include "llpanelplaces.h"
+#include "fsfloaterposestand.h"
+#include "fsfloaterteleporthistory.h"
 
 // Third party library includes
 #include <boost/algorithm/string.hpp>
@@ -146,7 +153,10 @@ static bool handleTerrainDetailChanged(const LLSD& newvalue)
 }
 
 
-static bool handleSetShaderChanged(const LLSD& newvalue)
+// <FS:Ansariel> Expose handleSetShaderChanged()
+//static bool handleSetShaderChanged(const LLSD& newvalue)
+bool handleSetShaderChanged(const LLSD& newvalue)
+// </FS:Ansariel>
 {
 	// changing shader level may invalidate existing cached bump maps, as the shader type determines the format of the bump map it expects - clear and repopulate the bump cache
 	gBumpImageList.destroyGL();
@@ -502,17 +512,17 @@ bool handleVoiceClientPrefsChanged(const LLSD& newvalue)
 // NaCl - Antispam Registry
 bool handleNaclAntiSpamGlobalQueueChanged(const LLSD& newvalue)
 {
-	NACLAntiSpamRegistry::setGlobalQueue(newvalue.asBoolean());
+	NACLAntiSpamRegistry::instance().setGlobalQueue(newvalue.asBoolean());
 	return true;
 }
 bool handleNaclAntiSpamTimeChanged(const LLSD& newvalue)
 {
-	NACLAntiSpamRegistry::setAllQueueTimes(newvalue.asInteger());
+	NACLAntiSpamRegistry::instance().setAllQueueTimes(newvalue.asInteger());
 	return true;
 }
 bool handleNaclAntiSpamAmountChanged(const LLSD& newvalue)
 {
-	NACLAntiSpamRegistry::setAllQueueAmounts(newvalue.asInteger());
+	NACLAntiSpamRegistry::instance().setAllQueueAmounts(newvalue.asInteger());
 	return true;
 }
 // NaCl End 
@@ -553,7 +563,10 @@ bool handleSettingF32Change(const LLSD& sdValue, F32* pValue)
 // ## Zi: Moved Avatar Z offset from RLVa to here
 bool handleAvatarZOffsetChanged(const LLSD& sdValue)
 {
-	gAgent.sendAgentSetAppearance();
+	if (isAgentAvatarValid())
+	{
+		gAgentAvatarp->computeBodySize();
+	}
 	return true;
 }
 // ## Zi: Moved Avatar Z offset from RLVa to here
@@ -564,6 +577,20 @@ bool handleForceShowGrid(const LLSD& newvalue)
 	return true;
 }
 
+bool handleLoginLocationChanged()
+{
+	/*
+	 * This connects the default preference setting to the state of the login
+	 * panel if it is displayed; if you open the preferences panel before
+	 * logging in, and change the default login location there, the login
+	 * panel immediately changes to match your new preference.
+	 */
+	std::string new_login_location = gSavedSettings.getString("LoginLocation");
+	LL_DEBUGS("AppInit")<<new_login_location<<LL_ENDL;
+	LLStartUp::setStartSLURL(LLSLURL(new_login_location));
+	return true;
+}
+
 bool handleSpellCheckChanged()
 {
 	if (gSavedSettings.getBOOL("SpellCheck"))
@@ -599,44 +626,23 @@ bool toggle_agent_pause(const LLSD& newvalue)
 // <FS:Zi> Is done inside XUI now, using visibility_control
 // bool toggle_show_navigation_panel(const LLSD& newvalue)
 // {
-// 	bool value = newvalue.asBoolean();
+	//bool value = newvalue.asBoolean();
 
-// 	//LLNavigationBar::getInstance()->setVisible(value);
-// 	LLNavigationBar::getInstance()->showNavigationPanel(value); // <FS:Ansariel> Re-enable separate toggle for navigation and favorites panel
-// //	gSavedSettings.setBOOL("ShowMiniLocationPanel", !value);
+	//LLNavigationBar::getInstance()->setVisible(value);
+	//gSavedSettings.setBOOL("ShowMiniLocationPanel", !value);
 
-// 	return true;
+	//return true;
 // }
 
-// <FS:Ansariel> Re-enable separate toggle for navigation and favorites panel
-// bool toggle_show_favorites_panel(const LLSD& newvalue)
-// {
-// 	LLNavigationBar::getInstance()->showFavoritesPanel(newvalue.asBoolean());
-// 	return true;
-// }
-// </FS:Ansariel>
-// </FS:Zi>
-
 // <FS:Zi> We don't have the mini location bar
 // bool toggle_show_mini_location_panel(const LLSD& newvalue)
 // {
-// 	bool value = newvalue.asBoolean();
-// 
-// 	LLPanelTopInfoBar::getInstance()->setVisible(value);
-// 	gSavedSettings.setBOOL("ShowNavbarNavigationPanel", !value);
-// 
-// 	return true;
-// }
-// </FS:Zi>
+	//bool value = newvalue.asBoolean();
 
-// <FS:Zi> Is done inside XUI now, using visibility_control
-// bool toggle_show_search_topbar(const LLSD& newvalue)
-// {
-// 	bool value = newvalue.asBoolean();
-// 	LLNavigationBar* navbar = LLNavigationBar::getInstance();
-// 	navbar->childSetVisible("search_combo_box",value);
-// 	return true;
-// }
+	//LLPanelTopInfoBar::getInstance()->setVisible(value);
+	//gSavedSettings.setBOOL("ShowNavbarNavigationPanel", !value);
+
+	//return true;
 // </FS:Zi>
 
 bool toggle_show_menubar_location_panel(const LLSD& newvalue)
@@ -691,6 +697,35 @@ static void handleNetMapDoubleClickActionChanged()
 }
 // </FS:Ansariel> Synchronize tooltips throughout instances
 
+// <FS:Ansariel> Clear places / teleport history search filter
+static void handleUseStandaloneTeleportHistoryFloaterChanged()
+{
+	LLFloaterSidePanelContainer* places = LLFloaterReg::findTypedInstance<LLFloaterSidePanelContainer>("places");
+	if (places)
+	{
+		places->findChild<LLPanelPlaces>("main_panel")->resetFilter();
+	}
+	FSFloaterTeleportHistory* tphistory = LLFloaterReg::findTypedInstance<FSFloaterTeleportHistory>("fs_teleporthistory");
+	if (tphistory)
+	{
+		tphistory->resetFilter();
+	}
+}
+// </FS:Ansariel> Clear places / teleport history search filter
+
+// <FS:CR> Posestand Ground Lock
+static void handleSetPoseStandLock(const LLSD& newvalue)
+{
+	FSFloaterPoseStand* pose_stand = LLFloaterReg::findTypedInstance<FSFloaterPoseStand>("fs_posestand");
+	if (pose_stand)
+	{
+		pose_stand->setLock(newvalue);
+		pose_stand->onCommitCombo();
+	}
+		
+}
+// </FS:CR> Posestand Ground Lock
+
 ////////////////////////////////////////////////////////////////////////////
 
 void settings_setup_listeners()
@@ -845,9 +880,6 @@ void settings_setup_listeners()
 	// <FS:Zi> We don't have the mini location bar
 	// gSavedSettings.getControl("ShowMiniLocationPanel")->getSignal()->connect(boost::bind(&toggle_show_mini_location_panel, _2));
 	// </FS: Zi>
-	// <FS:Zi> Is done inside XUI now, using visibility_control
-	// gSavedSettings.getControl("ShowSearchTopBar")->getSignal()->connect(boost::bind(&toggle_show_search_topbar, _2));
-	// </FS:Zi>
 	gSavedSettings.getControl("ShowMenuBarLocation")->getSignal()->connect(boost::bind(&toggle_show_menubar_location_panel, _2));
 	gSavedSettings.getControl("ShowObjectRenderingCost")->getSignal()->connect(boost::bind(&toggle_show_object_render_cost, _2));
 	gSavedSettings.getControl("UpdaterServiceSetting")->getSignal()->connect(boost::bind(&toggle_updater_service_active, _2));
@@ -855,9 +887,11 @@ void settings_setup_listeners()
 	gSavedSettings.getControl("RenderTransparentWater")->getSignal()->connect(boost::bind(&handleRenderTransparentWaterChanged, _2));
 	gSavedSettings.getControl("SpellCheck")->getSignal()->connect(boost::bind(&handleSpellCheckChanged));
 	gSavedSettings.getControl("SpellCheckDictionary")->getSignal()->connect(boost::bind(&handleSpellCheckChanged));
+	gSavedSettings.getControl("LoginLocation")->getSignal()->connect(boost::bind(&handleLoginLocationChanged));
 // [SL:KB] - Patch: UI-DndButtonCommit | Checked: 2011-06-19 (Catznip-2.6.0c) | Added: Catznip-2.6.0c
 	gSavedSettings.getControl("DragAndDropCommitDelay")->getSignal()->connect(boost::bind(&handleSettingF32Change, _2, &DELAY_DRAG_HOVER_COMMIT));
-	gSavedPerAccountSettings.getControl("AvatarZOffset")->getSignal()->connect(boost::bind(&handleAvatarZOffsetChanged, _2)); // ## Zi: Moved Avatar Z offset from RLVa to here
+	// <FS:CR> FIRE-9759 - Temporarily remove AvatarZOffset since it's broken
+	//gSavedPerAccountSettings.getControl("AvatarZOffset")->getSignal()->connect(boost::bind(&handleAvatarZOffsetChanged, _2)); // ## Zi: Moved Avatar Z offset from RLVa to here
 	gSavedSettings.getControl("FSUseV1Menus")->getSignal()->connect(boost::bind(&show_v1_menus));	// V1 menu system	-WoLf
 	// <FS:Zi> Is done inside XUI now, using visibility_control
 	// gSavedSettings.getControl("ShowNavbarFavoritesPanel")->getSignal()->connect(boost::bind(&toggle_show_favorites_panel, _2));
@@ -871,6 +905,15 @@ void settings_setup_listeners()
 
 	// <FS:Ansariel> Synchronize tooltips throughout instances
 	gSavedSettings.getControl("FSNetMapDoubleClickAction")->getSignal()->connect(boost::bind(&handleNetMapDoubleClickActionChanged));
+
+	// <FS:Ansariel> Clear places / teleport history search filter
+	gSavedSettings.getControl("FSUseStandaloneTeleportHistoryFloater")->getSignal()->connect(boost::bind(&handleUseStandaloneTeleportHistoryFloaterChanged));
+
+	// <FS:Ansariel> Tofu's SSR
+	gSavedSettings.getControl("FSRenderSSR")->getSignal()->connect(boost::bind(&handleSetShaderChanged, _2));
+	
+	// <FS:CR> Pose stand ground lock
+	gSavedSettings.getControl("FSPoseStandLock")->getSignal()->connect(boost::bind(&handleSetPoseStandLock, _2));
 }
 
 #if TEST_CACHED_CONTROL
