diff --git a/Source/indra/newview/llviewerdisplayname.cpp b/Source/indra/newview/llviewerdisplayname.cpp
index dae3f7a..2fe1e50 100644
--- a/Source/indra/newview/llviewerdisplayname.cpp
+++ b/Source/indra/newview/llviewerdisplayname.cpp
@@ -39,6 +39,8 @@
 #include "llhttpnode.h"
 #include "llnotificationsutil.h"
 #include "llui.h"					// getLanguage()
+
+#include "fsradar.h"
 #include "lggcontactsets.h"
 #include "llviewercontrol.h"
 
@@ -61,8 +63,10 @@ class LLSetDisplayNameResponder : public LLHTTPClient::Responder
 {
 public:
 	// only care about errors
-	/*virtual*/ void error(U32 status, const std::string& reason)
+	/*virtual*/ void errorWithContent(U32 status, const std::string& reason, const LLSD& content)
 	{
+		llwarns << "LLSetDisplayNameResponder error [status:"
+				<< status << "]: " << content << llendl;
 		LLViewerDisplayName::sSetDisplayNameSignal(false, "", LLSD());
 		LLViewerDisplayName::sSetDisplayNameSignal.disconnect_all_slots();
 	}
@@ -213,6 +217,16 @@ class LLDisplayNameUpdate : public LLHTTPNode
 		{
 			LLViewerDisplayName::sNameChangedSignal();
 		}
+		// <FS:Ansariel> Update name in radar
+		else
+		{
+			FSRadar* radar = FSRadar::getInstance();
+			if (radar)
+			{
+				radar->updateName(agent_id);
+			}
+		}
+		// </FS:Ansariel>
 	}
 };
 
