diff --git a/Source/indra/newview/llviewergesture.cpp b/Source/indra/newview/llviewergesture.cpp
index 1c20b77..0ed0109 100644
--- a/Source/indra/newview/llviewergesture.cpp
+++ b/Source/indra/newview/llviewergesture.cpp
@@ -42,7 +42,7 @@
 #include "llagent.h"
 // <FS:Zi> Remove floating chat bar
 // #include "llnearbychatbar.h"
-#include "llnearbychathub.h"
+#include "fsnearbychathub.h"
 // </FS:Zi>
 
 // Globals
@@ -135,7 +135,7 @@ void LLViewerGesture::doTrigger( BOOL send_chat )
 		// with the gesture animation.
 		// <FS:Zi> Remove floating chat bar
 		// LLNearbyChatBar::getInstance()->sendChatFromViewer(mOutputString, CHAT_TYPE_NORMAL, FALSE);
-		LLNearbyChat::instance().sendChatFromViewer(mOutputString, CHAT_TYPE_NORMAL, FALSE);
+		FSNearbyChat::instance().sendChatFromViewer(mOutputString, CHAT_TYPE_NORMAL, FALSE);
 	}
 }
 
