diff --git a/Source/indra/newview/llvoground.cpp b/Source/indra/newview/llvoground.cpp
index 6da5443..97b7418 100644
--- a/Source/indra/newview/llvoground.cpp
+++ b/Source/indra/newview/llvoground.cpp
@@ -49,18 +49,8 @@ LLVOGround::~LLVOGround()
 {
 }
 
-BOOL LLVOGround::idleUpdate(LLAgent &agent, LLWorld &world, const F64 &time)
+void LLVOGround::idleUpdate(LLAgent &agent, LLWorld &world, const F64 &time)
 {
- 	if (mDead || !(gPipeline.hasRenderType(LLPipeline::RENDER_TYPE_GROUND)))
-	{
-		return TRUE;
-	}
-	
-	/*if (mDrawable)
-	{
-		gPipeline.markRebuild(mDrawable, LLDrawable::REBUILD_VOLUME, TRUE);
-	}*/
-	return TRUE;
 }
 
 
