diff --git a/Source/indra/newview/llvograss.h b/Source/indra/newview/llvograss.h
index 00a59fa..27c38f1 100644
--- a/Source/indra/newview/llvograss.h
+++ b/Source/indra/newview/llvograss.h
@@ -73,11 +73,14 @@ public:
 	void plantBlades();
 
 	/*virtual*/ BOOL    isActive() const; // Whether this object needs to do an idleUpdate.
-	BOOL idleUpdate(LLAgent &agent, LLWorld &world, const F64 &time);
+	/*virtual*/ void idleUpdate(LLAgent &agent, LLWorld &world, const F64 &time);
 
 	/*virtual*/ BOOL lineSegmentIntersect(const LLVector3& start, const LLVector3& end, 
 										  S32 face = -1,                        // which face to check, -1 = ALL_SIDES
 										  BOOL pick_transparent = FALSE,
+// [SL:KB] - Patch: UI-PickRiggedAttachment | Checked: 2012-07-12 (Catznip-3.3)
+										  BOOL pick_rigged = FALSE,
+// [/SL:KB]
 										  S32* face_hit = NULL,                 // which face was hit
 										  LLVector3* intersection = NULL,       // return the intersection point
 										  LLVector2* tex_coord = NULL,          // return the texture coordinates of the intersection point
