diff --git a/Source/indra/newview/llwatchdog.h b/Source/indra/newview/llwatchdog.h
index fee3ec6..7b20d4c 100644
--- a/Source/indra/newview/llwatchdog.h
+++ b/Source/indra/newview/llwatchdog.h
@@ -61,15 +61,30 @@ public:
 	/* virtual */ void start() { start(""); }
 	/* virtual */ void stop();
 
-	void start(const std::string& state); 
+	// <FS:ND> Change from std::string to char const*, saving a lot of object construction/destruction per frame
+
+	// void start(const std::string& state); 
+	// void setTimeout(F32 d);
+	// void ping(const std::string& state);
+	// const std::string& getState() {return mPingState; }
+
+	void start( char const *state); 
 	void setTimeout(F32 d);
-	void ping(const std::string& state);
-	const std::string& getState() {return mPingState; }
+	void ping( char const *state );
+	std::string getState() {return mPingState; }
+
+	// </FS:ND>
 
 private:
 	LLTimer mTimer;
 	F32 mTimeout;
-	std::string mPingState;
+
+	// <FS:ND> Change from std::string to char const*, saving a lot of object construction/destruction per frame
+
+	// std::string mPingState;
+	char const *mPingState;
+
+	// </FS:ND>
 };
 
 class LLWatchdogTimerThread; // Defined in the cpp
