diff --git a/Source/indra/newview/pipeline.h b/Source/indra/newview/pipeline.h
index 2a2c9b5..e9ed442 100644
--- a/Source/indra/newview/pipeline.h
+++ b/Source/indra/newview/pipeline.h
@@ -78,7 +78,7 @@ void glh_set_current_modelview(const glh::matrix4f& mat);
 glh::matrix4f glh_get_current_projection();
 void glh_set_current_projection(glh::matrix4f& mat);
 glh::matrix4f gl_ortho(GLfloat left, GLfloat right, GLfloat bottom, GLfloat top, GLfloat znear, GLfloat zfar);
-glh::matrix4f gl_perspective(GLfloat fovy, GLfloat aspect, GLfloat zNear, GLfloat zFar);
+glh::matrix4f gl_perspective(GLfloat fovy, GLfloat aspect, GLfloat zNear, GLfloat zFar, bool render);
 glh::matrix4f gl_lookat(LLVector3 eye, LLVector3 center, LLVector3 up);
 
 extern LLFastTimer::DeclareTimer FTM_RENDER_GEOMETRY;
