diff --git a/PrepPush.bat b/PrepPush.bat
index 743c55c..0bb97fd 100644
--- a/PrepPush.bat
+++ b/PrepPush.bat
@@ -23,20 +23,21 @@ git add Source/*
 git commit -m "Latest from Firestorm Trunk"
 
 git checkout WorkingHead
-xcopy /d /y ..%src%*.cpp Source%src%
-xcopy /d /y ..%src%*.h Source%src%
-xcopy /d /y ..%settings%*.xml Source%settings%
-xcopy /d /y ..%messages%*.msg Source%messages%
+xcopy /y ..%src%*.cpp Source%src%
+xcopy /y ..%src%*.h Source%src%
+xcopy /y ..%settings%*.xml Source%settings%
+xcopy /y ..%messages%*.msg Source%messages%
 git add Source/*
 git commit -m "Latest from working directory"
 
 git checkout %head%
+
 for /f "delims=" %%l in ('git diff --name-only FirestormHead WorkingHead') do (
 	call:subroutine "%%l"
 )
 
+xcopy /s /c /e /h /i /r /y /exclude:.copyignore ..%messages%*.msg ..\build-vc100\newview\Release\app_settings\
 xcopy /s /c /d /e /h /i /r /y /exclude:.copyignore ..\build-vc100\newview\Release\* Bin\
-xcopy /s /c /d /e /h /i /r /y /exclude:.copyignore ..%messages%message_template.msg ..%settings%
 xcopy /s /c /d /e /h /i /r /y /exclude:.copyignore ..%src%* Bin\
 xcopy /s /c /d /e /h /i /r /y /exclude:.copyignore ..%settings%* Bin\app_settings\
 
