diff --git a/Source/indra/newview/utilitybar.cpp b/Source/indra/newview/utilitybar.cpp
index d6e887b..70214be 100644
--- a/Source/indra/newview/utilitybar.cpp
+++ b/Source/indra/newview/utilitybar.cpp
@@ -19,14 +19,19 @@
  * You should have received a copy of the GNU Lesser General Public
  * License along with this library; if not, write to the Free Software
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ * $/LicenseInfo$
  */
 
 #include "llviewerprecompiledheaders.h"
 
 #include "utilitybar.h"
 
-#include "llagent.h"
 #include "llbutton.h"
+#include "llfloaterreg.h"
+#include "llpanel.h"
+#include "lltabcontainer.h"
+
+#include "llagent.h"
 #include "llstatusbar.h"
 #include "llviewercontrol.h"
 #include "llviewermedia.h"
@@ -51,6 +56,7 @@ void UtilityBar::init()
 	mParcelMediaPlayButton = rootView->findChild<LLButton>("utility_parcel_media_button");
 	mTalkButton = rootView->findChild<LLButton>("utility_talk_button");
 	mAOInterfaceButton = rootView->findChild<LLButton>("show_ao_interface_button");
+	mVolumeControlsInterfaceButton = rootView->findChild<LLButton>("show_volume_controls_button");
 
 	if(mParcelStreamPlayButton)
 		mParcelStreamPlayButton->setCommitCallback(boost::bind(&UtilityBar::onParcelStreamClicked,this));
@@ -137,4 +143,19 @@ void UtilityBar::setAOInterfaceButtonExpanded(bool expanded)
 			mAOInterfaceButton->setImageOverlay("arrow_up.tga");
 		}
 	}
-}
\ No newline at end of file
+}
+
+void UtilityBar::setVolumeControlsButtonExpanded(bool expanded)
+{
+	if (mVolumeControlsInterfaceButton)
+	{
+		if (expanded)
+		{
+			mVolumeControlsInterfaceButton->setImageOverlay("arrow_down.tga");
+		}
+		else
+		{
+			mVolumeControlsInterfaceButton->setImageOverlay("arrow_up.tga");
+		}
+	}
+}
